{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b38ef9b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing the required modules\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.stats import anderson, wilcoxon, ttest_ind"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8d0818c",
   "metadata": {},
   "outputs": [],
   "source": [
    "home_dir = os.path.join(r'Z:/balint/numt/Michael/data/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "76590424",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>eg2_value</th>\n",
       "      <th>e_value</th>\n",
       "      <th>g_id</th>\n",
       "      <th>g_start</th>\n",
       "      <th>mt_start</th>\n",
       "      <th>g_length</th>\n",
       "      <th>mt_length</th>\n",
       "      <th>g_strand</th>\n",
       "      <th>mt_strand</th>\n",
       "      <th>g_size</th>\n",
       "      <th>g_sequence</th>\n",
       "      <th>mt_sequence</th>\n",
       "      <th>g_up_flanking</th>\n",
       "      <th>g_down_flanking</th>\n",
       "      <th>mt_up_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "      <th>mt_down_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3398</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1</td>\n",
       "      <td>114731432</td>\n",
       "      <td>6976</td>\n",
       "      <td>4680</td>\n",
       "      <td>4681</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>194850757</td>\n",
       "      <td>taGTAAAATTATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...</td>\n",
       "      <td>TAGTAAAACCATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...</td>\n",
       "      <td>GGGCAATTCAATATTCATAAATCAATAAACATCATAAATCACACCA...</td>\n",
       "      <td>TATATATATTTATATATATATGTATATATGTACATATATATAACTA...</td>\n",
       "      <td>GCTTTTAACAGACCGAAACTTAAATACAACCTTCTTTGATCCTGCA...</td>\n",
       "      <td>AACTGCTAATTCATGCTCCCATGTCTAACCCCATGGCTTTCTCAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3120</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GL019308</td>\n",
       "      <td>67160</td>\n",
       "      <td>883</td>\n",
       "      <td>4892</td>\n",
       "      <td>4945</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>88228</td>\n",
       "      <td>CAAGGCCATGAAGCATGCACACACTGCCCATCGCTCTCCTCAAACA...</td>\n",
       "      <td>CAAGGCCATGAAGCACGCACACACCGCCCGTCACCCTCCTCAAGTG...</td>\n",
       "      <td>AAAGCAATTAAATTTTTAACAAAATTTTTAGGCTTAATGAACTAGA...</td>\n",
       "      <td>ATGAAGGGGCAAAGAGTAATTTTACAGTCGAGAGGGCTAGCAGTCA...</td>\n",
       "      <td>GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...</td>\n",
       "      <td>TAGGGGCTATTAACTTTATTACAACTATTATTAATATGAAAGCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1888</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>GL019308</td>\n",
       "      <td>62527</td>\n",
       "      <td>12805</td>\n",
       "      <td>3439</td>\n",
       "      <td>3462</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>88228</td>\n",
       "      <td>AATTATTCATAGCCTCAATGATGAACAAGACATCCAAAAAACAGGA...</td>\n",
       "      <td>AATTATTCACAGCCTcaaTGATGAACAGGATATCCGAAAGATAGGC...</td>\n",
       "      <td>TTCTAATTACAGCACTCTACTCCCTCTACATTTTATCAACCAAAGA...</td>\n",
       "      <td>ACGTATACATGAGTTGTTCTTACATAATTGAATATACCTGCAATGC...</td>\n",
       "      <td>TTATTATTTTAGTCCTCCCTATTGTGGCCTCTTTCACTAACATTTT...</td>\n",
       "      <td>GTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGTACAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1824</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>AAGW02081594</td>\n",
       "      <td>14016</td>\n",
       "      <td>6061</td>\n",
       "      <td>3040</td>\n",
       "      <td>3073</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>24576</td>\n",
       "      <td>TTCTTTGGACACCCTGAGGTATATATTCTTATCCTTCCAGGATTCG...</td>\n",
       "      <td>TTTTTCGGGCACCCCGAAGTATATATTCTTATTCTTCCAGGATTTG...</td>\n",
       "      <td>TGCCACACCTCCCATGGGATAAAGAGGAAAGGGTCTGGTGCCATGC...</td>\n",
       "      <td>TGAAATAGCTATCACTATTCAGGTATGCCACTATTCTTCTAGGCAT...</td>\n",
       "      <td>CATCTCCTGAGTGCAAACCAGGTGCTTTAATTAAGCTAAATCCTCC...</td>\n",
       "      <td>ATTCTCCTAGGTATCTATTTTACCCTACTTCAAGCATCGGAGTATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1344</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>AAGW02081594</td>\n",
       "      <td>20039</td>\n",
       "      <td>11341</td>\n",
       "      <td>2339</td>\n",
       "      <td>2362</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>24576</td>\n",
       "      <td>ATAGCATCCTTTTCATGATCTAATCTCACTATTATTTTAATAGGGA...</td>\n",
       "      <td>ATGGCCTCCTTCTCATGATCTAACCTCACAATCATCCTGATAGGGA...</td>\n",
       "      <td>TCTTTCTCATTGTCCACTGTGCCTGTCAAAAAATAAAAAGAAAAAG...</td>\n",
       "      <td>GGCTGGTGCTGCAGCTCAGTAGGCTAATCTTCCACCTAGCGGCGTG...</td>\n",
       "      <td>CCTAAACTTCTCAACCACATTTTTTTCAGATGCCTTATCCACTCCT...</td>\n",
       "      <td>ATCCTACCTCGTCCCCTTCAAAAACCACCCAATCCCCCATATTTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>80</td>\n",
       "      <td>2.800000e-21</td>\n",
       "      <td>5.000000e-25</td>\n",
       "      <td>GL019047</td>\n",
       "      <td>205885</td>\n",
       "      <td>9539</td>\n",
       "      <td>243</td>\n",
       "      <td>247</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>247428</td>\n",
       "      <td>GATCAGTTTTCAAAGTGCTTTTATGGTACTGA----AGTACAATAG...</td>\n",
       "      <td>GATCAGTTTTCGAAGTGTTTAAGTGGAACTATTTCGAGGACAATAG...</td>\n",
       "      <td>TTAATTAGATGGCTAAGATTGTGTGTTGAATAAATAGAGTAGTAAT...</td>\n",
       "      <td>AGTTAGGGAGGAGTTGTTAATTTTGTCTGTTATGCATAGAGCTCAT...</td>\n",
       "      <td>CACCTACTATGTAGTAGCTCACTTCCACTATGTCCTATCTATGGGG...</td>\n",
       "      <td>CCTATCTATGATTTAGACTCGCTATGAAGCTAAACAGCGCTAGCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>77</td>\n",
       "      <td>7.500000e-20</td>\n",
       "      <td>1.300000e-23</td>\n",
       "      <td>AAGW02081404</td>\n",
       "      <td>17858</td>\n",
       "      <td>882</td>\n",
       "      <td>136</td>\n",
       "      <td>140</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>27691</td>\n",
       "      <td>GTACGTgcgtacgtgcgtacgggtgtacgggtgtacgggtgtacgg...</td>\n",
       "      <td>gtgcgtacgggtgtacgtgcgtgcgtacgggtgtacgtgcgtgcgt...</td>\n",
       "      <td>TTGTAAAGGTGCGTTTTAAATTTCAAAATTTAGGTGCCCTGAGCTG...</td>\n",
       "      <td>GTACGTGCGTACGTGCGTACGGGTGTACGGGTGTACGTGCGTGCGT...</td>\n",
       "      <td>TGGACAAGTAGCATCTGTCCTCTACTTCACCACCATCCTTATTCTT...</td>\n",
       "      <td>GCACGTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>150</th>\n",
       "      <td>75</td>\n",
       "      <td>6.700000e-19</td>\n",
       "      <td>1.200000e-22</td>\n",
       "      <td>GL019311</td>\n",
       "      <td>22172</td>\n",
       "      <td>5436</td>\n",
       "      <td>218</td>\n",
       "      <td>228</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>80506</td>\n",
       "      <td>ATAATTAGTCAAGTTATAGAGGAAGAAAATAG----ATTAGTTTTA...</td>\n",
       "      <td>ATAATTGATACAGCTACAGAGGTGGAGAATAAGTTAATTACTTTTA...</td>\n",
       "      <td>CCCGCGGGTCGTGCGCCAGGTGAGGGCGCAGGTTCGGCCCGGGCCC...</td>\n",
       "      <td>GTTTAAAATGTATGTGCCCTGGGGCTGGCGCGCTGGTGCAGTAGGT...</td>\n",
       "      <td>TCTTATCATTATTACACGATGAGGTAATCAAACAGAGCGACTCAAC...</td>\n",
       "      <td>TATTTTAGTCCTCCCTATTGTGGCCTCTTTCACTAACATTTTTAAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151</th>\n",
       "      <td>74</td>\n",
       "      <td>2.000000e-18</td>\n",
       "      <td>3.600000e-22</td>\n",
       "      <td>21</td>\n",
       "      <td>11379492</td>\n",
       "      <td>15954</td>\n",
       "      <td>321</td>\n",
       "      <td>320</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>15578276</td>\n",
       "      <td>CTAATTCATTACGTAAAAAGATAAAAGGGTTAATCTTTGCTTTTTA...</td>\n",
       "      <td>CTAATTCATTATGCAAAAGG-TAAGAGAAGTAATCTCTGCTTTTTT...</td>\n",
       "      <td>ATGACAGCCACAGTGAGCACTAAAGTTCTGTTGCACAGAATTGTTT...</td>\n",
       "      <td>GCATGTTTAAGTTTATGATTTAAATTTAAATTAGCTTAGGTTGGCT...</td>\n",
       "      <td>TGGACTAGAAATCTCTAGTCTAGGCTAAAGTTAATGTAGCTTAACA...</td>\n",
       "      <td>CTAGAAAATCCTTAGCAAAAAGAATTTTAGTTAAGAACCCCGAAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>73</td>\n",
       "      <td>6.000000e-18</td>\n",
       "      <td>1.100000e-21</td>\n",
       "      <td>GL018744</td>\n",
       "      <td>673556</td>\n",
       "      <td>4478</td>\n",
       "      <td>319</td>\n",
       "      <td>330</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>2220855</td>\n",
       "      <td>GTGCCTGCAGATGTAGAGAAAAACTGAGTGTGGCTGATTAACGCCA...</td>\n",
       "      <td>GTGGGTGCAGATGTGGAGAAATGCCAGGTGTGGTTGGTTAATTCCG...</td>\n",
       "      <td>GTAGCTCTCACTTTCAAATAAATAAAATAAATCTTTTAAAAAATCT...</td>\n",
       "      <td>ACTACTGCTGTTAGGAGGAGCCGCGTTAATGGGAATGAATCGGGGT...</td>\n",
       "      <td>CAACGAGGCAAATTCACGTATCACACAAACAACATTTCCCCTACAT...</td>\n",
       "      <td>GCCTTCTTCAAAGCTATGCTCTTTCTCTGCTCCGGCTCAATTATTC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     score     eg2_value       e_value          g_id    g_start  mt_start  \\\n",
       "1     3398  0.000000e+00  0.000000e+00             1  114731432      6976   \n",
       "2     3120  0.000000e+00  0.000000e+00      GL019308      67160       883   \n",
       "3     1888  0.000000e+00  0.000000e+00      GL019308      62527     12805   \n",
       "4     1824  0.000000e+00  0.000000e+00  AAGW02081594      14016      6061   \n",
       "5     1344  0.000000e+00  0.000000e+00  AAGW02081594      20039     11341   \n",
       "..     ...           ...           ...           ...        ...       ...   \n",
       "148     80  2.800000e-21  5.000000e-25      GL019047     205885      9539   \n",
       "149     77  7.500000e-20  1.300000e-23  AAGW02081404      17858       882   \n",
       "150     75  6.700000e-19  1.200000e-22      GL019311      22172      5436   \n",
       "151     74  2.000000e-18  3.600000e-22            21   11379492     15954   \n",
       "152     73  6.000000e-18  1.100000e-21      GL018744     673556      4478   \n",
       "\n",
       "     g_length  mt_length g_strand mt_strand     g_size  \\\n",
       "1        4680       4681        +         +  194850757   \n",
       "2        4892       4945        +         +      88228   \n",
       "3        3439       3462        +         +      88228   \n",
       "4        3040       3073        +         +      24576   \n",
       "5        2339       2362        +         +      24576   \n",
       "..        ...        ...      ...       ...        ...   \n",
       "148       243        247        +         -     247428   \n",
       "149       136        140        +         -      27691   \n",
       "150       218        228        +         -      80506   \n",
       "151       321        320        +         -   15578276   \n",
       "152       319        330        +         -    2220855   \n",
       "\n",
       "                                            g_sequence  \\\n",
       "1    taGTAAAATTATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...   \n",
       "2    CAAGGCCATGAAGCATGCACACACTGCCCATCGCTCTCCTCAAACA...   \n",
       "3    AATTATTCATAGCCTCAATGATGAACAAGACATCCAAAAAACAGGA...   \n",
       "4    TTCTTTGGACACCCTGAGGTATATATTCTTATCCTTCCAGGATTCG...   \n",
       "5    ATAGCATCCTTTTCATGATCTAATCTCACTATTATTTTAATAGGGA...   \n",
       "..                                                 ...   \n",
       "148  GATCAGTTTTCAAAGTGCTTTTATGGTACTGA----AGTACAATAG...   \n",
       "149  GTACGTgcgtacgtgcgtacgggtgtacgggtgtacgggtgtacgg...   \n",
       "150  ATAATTAGTCAAGTTATAGAGGAAGAAAATAG----ATTAGTTTTA...   \n",
       "151  CTAATTCATTACGTAAAAAGATAAAAGGGTTAATCTTTGCTTTTTA...   \n",
       "152  GTGCCTGCAGATGTAGAGAAAAACTGAGTGTGGCTGATTAACGCCA...   \n",
       "\n",
       "                                           mt_sequence  \\\n",
       "1    TAGTAAAACCATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...   \n",
       "2    CAAGGCCATGAAGCACGCACACACCGCCCGTCACCCTCCTCAAGTG...   \n",
       "3    AATTATTCACAGCCTcaaTGATGAACAGGATATCCGAAAGATAGGC...   \n",
       "4    TTTTTCGGGCACCCCGAAGTATATATTCTTATTCTTCCAGGATTTG...   \n",
       "5    ATGGCCTCCTTCTCATGATCTAACCTCACAATCATCCTGATAGGGA...   \n",
       "..                                                 ...   \n",
       "148  GATCAGTTTTCGAAGTGTTTAAGTGGAACTATTTCGAGGACAATAG...   \n",
       "149  gtgcgtacgggtgtacgtgcgtgcgtacgggtgtacgtgcgtgcgt...   \n",
       "150  ATAATTGATACAGCTACAGAGGTGGAGAATAAGTTAATTACTTTTA...   \n",
       "151  CTAATTCATTATGCAAAAGG-TAAGAGAAGTAATCTCTGCTTTTTT...   \n",
       "152  GTGGGTGCAGATGTGGAGAAATGCCAGGTGTGGTTGGTTAATTCCG...   \n",
       "\n",
       "                                         g_up_flanking  \\\n",
       "1    GGGCAATTCAATATTCATAAATCAATAAACATCATAAATCACACCA...   \n",
       "2    AAAGCAATTAAATTTTTAACAAAATTTTTAGGCTTAATGAACTAGA...   \n",
       "3    TTCTAATTACAGCACTCTACTCCCTCTACATTTTATCAACCAAAGA...   \n",
       "4    TGCCACACCTCCCATGGGATAAAGAGGAAAGGGTCTGGTGCCATGC...   \n",
       "5    TCTTTCTCATTGTCCACTGTGCCTGTCAAAAAATAAAAAGAAAAAG...   \n",
       "..                                                 ...   \n",
       "148  TTAATTAGATGGCTAAGATTGTGTGTTGAATAAATAGAGTAGTAAT...   \n",
       "149  TTGTAAAGGTGCGTTTTAAATTTCAAAATTTAGGTGCCCTGAGCTG...   \n",
       "150  CCCGCGGGTCGTGCGCCAGGTGAGGGCGCAGGTTCGGCCCGGGCCC...   \n",
       "151  ATGACAGCCACAGTGAGCACTAAAGTTCTGTTGCACAGAATTGTTT...   \n",
       "152  GTAGCTCTCACTTTCAAATAAATAAAATAAATCTTTTAAAAAATCT...   \n",
       "\n",
       "                                       g_down_flanking  \\\n",
       "1    TATATATATTTATATATATATGTATATATGTACATATATATAACTA...   \n",
       "2    ATGAAGGGGCAAAGAGTAATTTTACAGTCGAGAGGGCTAGCAGTCA...   \n",
       "3    ACGTATACATGAGTTGTTCTTACATAATTGAATATACCTGCAATGC...   \n",
       "4    TGAAATAGCTATCACTATTCAGGTATGCCACTATTCTTCTAGGCAT...   \n",
       "5    GGCTGGTGCTGCAGCTCAGTAGGCTAATCTTCCACCTAGCGGCGTG...   \n",
       "..                                                 ...   \n",
       "148  AGTTAGGGAGGAGTTGTTAATTTTGTCTGTTATGCATAGAGCTCAT...   \n",
       "149  GTACGTGCGTACGTGCGTACGGGTGTACGGGTGTACGTGCGTGCGT...   \n",
       "150  GTTTAAAATGTATGTGCCCTGGGGCTGGCGCGCTGGTGCAGTAGGT...   \n",
       "151  GCATGTTTAAGTTTATGATTTAAATTTAAATTAGCTTAGGTTGGCT...   \n",
       "152  ACTACTGCTGTTAGGAGGAGCCGCGTTAATGGGAATGAATCGGGGT...   \n",
       "\n",
       "    mt_up_flanking (if strand is negative, the coordinates are corrected)  \\\n",
       "1    GCTTTTAACAGACCGAAACTTAAATACAACCTTCTTTGATCCTGCA...                      \n",
       "2    GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...                      \n",
       "3    TTATTATTTTAGTCCTCCCTATTGTGGCCTCTTTCACTAACATTTT...                      \n",
       "4    CATCTCCTGAGTGCAAACCAGGTGCTTTAATTAAGCTAAATCCTCC...                      \n",
       "5    CCTAAACTTCTCAACCACATTTTTTTCAGATGCCTTATCCACTCCT...                      \n",
       "..                                                 ...                      \n",
       "148  CACCTACTATGTAGTAGCTCACTTCCACTATGTCCTATCTATGGGG...                      \n",
       "149  TGGACAAGTAGCATCTGTCCTCTACTTCACCACCATCCTTATTCTT...                      \n",
       "150  TCTTATCATTATTACACGATGAGGTAATCAAACAGAGCGACTCAAC...                      \n",
       "151  TGGACTAGAAATCTCTAGTCTAGGCTAAAGTTAATGTAGCTTAACA...                      \n",
       "152  CAACGAGGCAAATTCACGTATCACACAAACAACATTTCCCCTACAT...                      \n",
       "\n",
       "    mt_down_flanking (if strand is negative, the coordinates are corrected)  \n",
       "1    AACTGCTAATTCATGCTCCCATGTCTAACCCCATGGCTTTCTCAAC...                       \n",
       "2    TAGGGGCTATTAACTTTATTACAACTATTATTAATATGAAAGCCCC...                       \n",
       "3    GTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGTACAC...                       \n",
       "4    ATTCTCCTAGGTATCTATTTTACCCTACTTCAAGCATCGGAGTATT...                       \n",
       "5    ATCCTACCTCGTCCCCTTCAAAAACCACCCAATCCCCCATATTTTT...                       \n",
       "..                                                 ...                       \n",
       "148  CCTATCTATGATTTAGACTCGCTATGAAGCTAAACAGCGCTAGCCT...                       \n",
       "149  GCACGTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGT...                       \n",
       "150  TATTTTAGTCCTCCCTATTGTGGCCTCTTTCACTAACATTTTTAAT...                       \n",
       "151  CTAGAAAATCCTTAGCAAAAAGAATTTTAGTTAAGAACCCCGAAAC...                       \n",
       "152  GCCTTCTTCAAAGCTATGCTCTTTCTCTGCTCCGGCTCAATTATTC...                       \n",
       "\n",
       "[152 rows x 17 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numt_array = pd.read_csv(home_dir + r'numt_master_array_with_1000_flanking.csv')\n",
    "numt_array = numt_array.drop(0)\n",
    "numt_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b4cb7f4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1                 1\n",
       "2          GL019308\n",
       "3          GL019308\n",
       "4      AAGW02081594\n",
       "5      AAGW02081594\n",
       "           ...     \n",
       "148        GL019047\n",
       "149    AAGW02081404\n",
       "150        GL019311\n",
       "151              21\n",
       "152        GL018744\n",
       "Name: g_id, Length: 152, dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "g_ids = numt_array['g_id']\n",
    "g_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a0a242dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g_id\n",
       "1               194850757\n",
       "GL019308            88228\n",
       "GL019308            88228\n",
       "AAGW02081594        24576\n",
       "AAGW02081594        24576\n",
       "                  ...    \n",
       "GL019047           247428\n",
       "AAGW02081404        27691\n",
       "GL019311            80506\n",
       "21               15578276\n",
       "GL018744          2220855\n",
       "Name: g_size, Length: 152, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chromosome_length = numt_array['g_size']\n",
    "chromosome_length.index = g_ids\n",
    "chromosome_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c3b9b534",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g_id\n",
       "1               114726432\n",
       "GL019308            62160\n",
       "GL019308            57527\n",
       "AAGW02081594         9016\n",
       "AAGW02081594        15039\n",
       "                  ...    \n",
       "GL019047           200885\n",
       "AAGW02081404        12858\n",
       "GL019311            17172\n",
       "21               11374492\n",
       "GL018744           668556\n",
       "Name: g_start, Length: 152, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the start positions of the upstream flankings\n",
    "u_flanking_start = numt_array['g_start'].apply(lambda start : start - 5000)\n",
    "u_flanking_start[u_flanking_start < 0] = 1\n",
    "u_flanking_start.index = g_ids\n",
    "u_flanking_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c166fc73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g_id\n",
       "1               114731432\n",
       "GL019308            67160\n",
       "GL019308            62527\n",
       "AAGW02081594        14016\n",
       "AAGW02081594        20039\n",
       "                  ...    \n",
       "GL019047           205885\n",
       "AAGW02081404        17858\n",
       "GL019311            22172\n",
       "21               11379492\n",
       "GL018744           673556\n",
       "Name: g_start, Length: 152, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the end positions of the upstream flankings (which are the genomic_starts of the numts)\n",
    "u_flanking_end = numt_array['g_start']\n",
    "u_flanking_end.index = g_ids\n",
    "u_flanking_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "704da38e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g_id\n",
       "1               114736112\n",
       "GL019308            72052\n",
       "GL019308            65966\n",
       "AAGW02081594        17056\n",
       "AAGW02081594        22378\n",
       "                  ...    \n",
       "GL019047           206128\n",
       "AAGW02081404        17994\n",
       "GL019311            22390\n",
       "21               11379813\n",
       "GL018744           673875\n",
       "Length: 152, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#start position of downstream_flanking (which is equal to the ends of the numts)\n",
    "numt_array['g_length'].index = g_ids\n",
    "d_flanking_start = numt_array['g_start'] + numt_array['g_length']\n",
    "d_flanking_start.index = g_ids\n",
    "d_flanking_start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "86b43fd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g_id\n",
       "1               114741112\n",
       "GL019308            77052\n",
       "GL019308            70966\n",
       "AAGW02081594        22056\n",
       "AAGW02081594        27378\n",
       "                  ...    \n",
       "GL019047           211128\n",
       "AAGW02081404        22994\n",
       "GL019311            27390\n",
       "21               11384813\n",
       "GL018744           678875\n",
       "Length: 152, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#end position of downstream_flanking\n",
    "numt_ends = numt_array['g_start'] + numt_array['g_length']\n",
    "d_flanking_end = numt_ends.apply(lambda end : end + 5000)\n",
    "d_flanking_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "788b9987",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pytho\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:3: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  This is separate from the ipykernel package so we can avoid doing imports until\n",
      "C:\\Users\\pytho\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:4: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.\n",
      "  after removing the cwd from sys.path.\n"
     ]
    }
   ],
   "source": [
    "#get the regions\n",
    "indices = np.arange(0,len(u_flanking_start),1)\n",
    "upstream_flankings = pd.Series(index = indices)\n",
    "downstream_flankings = pd.Series(index = indices)\n",
    "for index, u_start in enumerate(u_flanking_start):\n",
    "    upstream_rm_item = (str(g_ids[index + 1]) + ':' +str(u_start) + '-' + str(u_flanking_end[index]))\n",
    "    upstream_flankings[index] = upstream_rm_item\n",
    "    downstream_rm_item = ''\n",
    "    if chromosome_length[index] < d_flanking_end[index]:\n",
    "        downstream_rm_item = (str(g_ids[index + 1]) + ':' +str(d_flanking_start[index]) +\n",
    "                          '-' + str(chromosome_length[index]))\n",
    "    else:\n",
    "        downstream_rm_item = (str(g_ids[index + 1]) + ':' +str(d_flanking_start[index]) +\n",
    "                              '-' + str(d_flanking_end[index]))\n",
    "    downstream_flankings[index] = downstream_rm_item"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "d147922e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         1:114726432-114731432\n",
       "1          GL019308:62160-67160\n",
       "2          GL019308:57527-62527\n",
       "3       AAGW02081594:9016-14016\n",
       "4      AAGW02081594:15039-20039\n",
       "                 ...           \n",
       "147      GL019047:200885-205885\n",
       "148    AAGW02081404:12858-17858\n",
       "149        GL019311:17172-22172\n",
       "150        21:11374492-11379492\n",
       "151      GL018744:668556-673556\n",
       "Length: 152, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "upstream_flankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e94166b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0         1:114736112-114741112\n",
       "1          GL019308:72052-77052\n",
       "2          GL019308:65966-70966\n",
       "3      AAGW02081594:17056-22056\n",
       "4      AAGW02081594:22378-24576\n",
       "                 ...           \n",
       "147      GL019047:206128-211128\n",
       "148    AAGW02081404:17994-22994\n",
       "149        GL019311:22390-27390\n",
       "150        21:11379813-11384813\n",
       "151      GL018744:673875-678875\n",
       "Length: 152, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "downstream_flankings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "22775f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writing output\n",
    "with open(home_dir + 'upstream_flankings.bed','w') as up_out, open(home_dir + 'downstream_flankings.bed','w') as down_out:\n",
    "    upstream_flankings.apply(lambda line : up_out.write(line + '\\n'))\n",
    "    downstream_flankings.apply(lambda line : down_out.write(line + '\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "405c37d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGW02081404.fasta\n",
      "MT.fasta\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1               AGGACCGCACAACCTTGCAGCTCAGCGACTCGTGGGGTCACACACA...\n",
       "11              CTCACACACCCACGCGGCCTCACACACCCACGAGACCTCACACACC...\n",
       "12              CTGAGTTCCTGGCTTCTGGCTTTGGCCTGGCTCAACCCCAGCTGCA...\n",
       "13              TCATAATCCTAACCATAACCAAGGATTCAACCATAAATTTAAGACT...\n",
       "14              AAATGAAGTCAGATATGGTAATCGATACTGAACTGAGGCACGGATT...\n",
       "15              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "16              CAGAATGAGATTTCTGCACTACAGACAAGAGTCTTTTGTCGTGCCT...\n",
       "17              GGTGTGGTATATGTGGTGTGTATGGTATGTGTGTGGTACTGTATGG...\n",
       "18              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "19              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "2               GAGTTATAAAAGAGAAGGAGAGGCAGAGAGAGAGAGAGCAGTATGA...\n",
       "20              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "21              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "3               AAGCAGGCCAGTCCACTGTGTCACTTACAGACACAGCTGACACTGA...\n",
       "4               AACCAACGAGCTATGCACTTATGATATATGTGCTTTTTTTTTTTTT...\n",
       "7               ACACACACAACACACACGACATAACAACACATACAACACACAAATA...\n",
       "9               CGGGGACTCCCACTAGAGGAACCAGGTGATGTCAGTTCCCAGAATC...\n",
       "AAGW02079688    TATTCCCCAACCTCTGATAACTACCATTCTACCGTGTCTACTTCTA...\n",
       "AAGW02081404                                                     \n",
       "AAGW02081594    TTATCTTTATTTCATCCATGATCAAATATTTCACCCACGATGATTA...\n",
       "AAGW02083714    TCCCCTAACTGCATGAGTCGGTTCCGAATTCACTCTCATTGTGAGT...\n",
       "GL018700        CCAGATGGATACATGAAGTATGCAGTTGCCTATGTAGTTTAGAACT...\n",
       "GL018705        ATAATGATCAGATGATAATGGTGGTGATGATGGTGAGGATGACTCT...\n",
       "GL018706        AGTTTGTTCCCACACAGAAACATAAAGTGAAAAATAATAGATGATT...\n",
       "GL018741        TAAAACTAAGAGTCATTCTAATCTTTCCTCTATATCTAATCCCAAC...\n",
       "GL018744        GCTGCATTTTATTTTGTAGACAGGCATATTTTCTTTTCCAATTGCT...\n",
       "GL018748        AATGGCCGCAACAGCGGGACCCTGGGCAATCTGAAGCCAGGAGTCA...\n",
       "GL018786        ACGCCCCGTCCAGCCAGACTGCCACTCCGCTGGCGGAACCCCCGAG...\n",
       "GL018817        ACCTAATTACAGGGCACCGGTCTATTTAAAAGATATGTTAAGGGAC...\n",
       "GL018826        ACATTATAAAACATAGAGACAGTGGGTAGTGTCCCACATACATGTG...\n",
       "GL018862        CAGTCAGTCATATTCATGGCTGTTAGCCTAGTATGTGAGATTGTCT...\n",
       "GL018925        AAAGTCTCAATACCTACCTAAGTCAGTTCCCAAGTCCTTTAGATTG...\n",
       "GL018992        GAGGTGGTTAGAAATAGAATCCTGACTATTGTCTCCAGTTTCTCTA...\n",
       "GL019010        CCCGCGGTCAGTAGCCGTCACTGGTGCTGCCACTATTGAGGCTGTC...\n",
       "GL019047        CTCCCCTTCCATTCAGTCCAGATCATTTTCGATTCTCTTTATATAC...\n",
       "GL019056        TACAGCATTGTACATGTTGTAACTCCCATCACTCATGGTCATATCA...\n",
       "GL019086        TCTCTGGATGCGACCTTTTAGTCATTCTGGAACAAAGTCTGACCTG...\n",
       "GL019149        CCAGATATCCATCAACTGATGACCGGATAAAGAAATCCTGGCATAC...\n",
       "GL019235        TAGCTGATGGGTATCAGTGGTTCTTGCCAGGAACCCCAGTGACCTC...\n",
       "GL019271        TGCTGCCTGTGATAGAAGTCAGTTTGTAGATCACCTTGAACCCTAC...\n",
       "GL019308        AGCATTTCCAGCATGCAGTCTCTCCCATGAAGGTGGAGAAGGAACT...\n",
       "GL019311        CCAGATCCTTGGGCCCTGCATCCACATGGGAGACCAGGAGGAAACA...\n",
       "GL019326        GAGACTGGGAAGAAGCTCCTGGCTCCTGGCTTTGGATCAGCGCAGC...\n",
       "GL019383        CTAAGCCAGAGACTTAGTCAAAAGACAAAAGCCATCAGAGGAAAAA...\n",
       "GL019465        ATACTTGGCTCCCCACCCAGGAGATCGCAGCTTGAGGCATCACCCC...\n",
       "GL019525        TTTTAATGATGATGAAATCAGATCAGACCTATTGTCATGTTTAATC...\n",
       "MT                                                               \n",
       "dtype: object"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numtless_sequences = numtless_ids.apply(get_numtless_sequences)\n",
    "numtless_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "20f63571",
   "metadata": {},
   "outputs": [],
   "source": [
    "numtless_sequences = numtless_sequences.drop(labels = ['AAGW02081404', 'MT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "83fbf5c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1               AGGACCGCACAACCTTGCAGCTCAGCGACTCGTGGGGTCACACACA...\n",
       "11              CTCACACACCCACGCGGCCTCACACACCCACGAGACCTCACACACC...\n",
       "12              CTGAGTTCCTGGCTTCTGGCTTTGGCCTGGCTCAACCCCAGCTGCA...\n",
       "13              TCATAATCCTAACCATAACCAAGGATTCAACCATAAATTTAAGACT...\n",
       "14              AAATGAAGTCAGATATGGTAATCGATACTGAACTGAGGCACGGATT...\n",
       "15              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "16              CAGAATGAGATTTCTGCACTACAGACAAGAGTCTTTTGTCGTGCCT...\n",
       "17              GGTGTGGTATATGTGGTGTGTATGGTATGTGTGTGGTACTGTATGG...\n",
       "18              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "19              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "2               GAGTTATAAAAGAGAAGGAGAGGCAGAGAGAGAGAGAGCAGTATGA...\n",
       "20              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "21              NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...\n",
       "3               AAGCAGGCCAGTCCACTGTGTCACTTACAGACACAGCTGACACTGA...\n",
       "4               AACCAACGAGCTATGCACTTATGATATATGTGCTTTTTTTTTTTTT...\n",
       "7               ACACACACAACACACACGACATAACAACACATACAACACACAAATA...\n",
       "9               CGGGGACTCCCACTAGAGGAACCAGGTGATGTCAGTTCCCAGAATC...\n",
       "AAGW02079688    TATTCCCCAACCTCTGATAACTACCATTCTACCGTGTCTACTTCTA...\n",
       "AAGW02081594    TTATCTTTATTTCATCCATGATCAAATATTTCACCCACGATGATTA...\n",
       "AAGW02083714    TCCCCTAACTGCATGAGTCGGTTCCGAATTCACTCTCATTGTGAGT...\n",
       "GL018700        CCAGATGGATACATGAAGTATGCAGTTGCCTATGTAGTTTAGAACT...\n",
       "GL018705        ATAATGATCAGATGATAATGGTGGTGATGATGGTGAGGATGACTCT...\n",
       "GL018706        AGTTTGTTCCCACACAGAAACATAAAGTGAAAAATAATAGATGATT...\n",
       "GL018741        TAAAACTAAGAGTCATTCTAATCTTTCCTCTATATCTAATCCCAAC...\n",
       "GL018744        GCTGCATTTTATTTTGTAGACAGGCATATTTTCTTTTCCAATTGCT...\n",
       "GL018748        AATGGCCGCAACAGCGGGACCCTGGGCAATCTGAAGCCAGGAGTCA...\n",
       "GL018786        ACGCCCCGTCCAGCCAGACTGCCACTCCGCTGGCGGAACCCCCGAG...\n",
       "GL018817        ACCTAATTACAGGGCACCGGTCTATTTAAAAGATATGTTAAGGGAC...\n",
       "GL018826        ACATTATAAAACATAGAGACAGTGGGTAGTGTCCCACATACATGTG...\n",
       "GL018862        CAGTCAGTCATATTCATGGCTGTTAGCCTAGTATGTGAGATTGTCT...\n",
       "GL018925        AAAGTCTCAATACCTACCTAAGTCAGTTCCCAAGTCCTTTAGATTG...\n",
       "GL018992        GAGGTGGTTAGAAATAGAATCCTGACTATTGTCTCCAGTTTCTCTA...\n",
       "GL019010        CCCGCGGTCAGTAGCCGTCACTGGTGCTGCCACTATTGAGGCTGTC...\n",
       "GL019047        CTCCCCTTCCATTCAGTCCAGATCATTTTCGATTCTCTTTATATAC...\n",
       "GL019056        TACAGCATTGTACATGTTGTAACTCCCATCACTCATGGTCATATCA...\n",
       "GL019086        TCTCTGGATGCGACCTTTTAGTCATTCTGGAACAAAGTCTGACCTG...\n",
       "GL019149        CCAGATATCCATCAACTGATGACCGGATAAAGAAATCCTGGCATAC...\n",
       "GL019235        TAGCTGATGGGTATCAGTGGTTCTTGCCAGGAACCCCAGTGACCTC...\n",
       "GL019271        TGCTGCCTGTGATAGAAGTCAGTTTGTAGATCACCTTGAACCCTAC...\n",
       "GL019308        AGCATTTCCAGCATGCAGTCTCTCCCATGAAGGTGGAGAAGGAACT...\n",
       "GL019311        CCAGATCCTTGGGCCCTGCATCCACATGGGAGACCAGGAGGAAACA...\n",
       "GL019326        GAGACTGGGAAGAAGCTCCTGGCTCCTGGCTTTGGATCAGCGCAGC...\n",
       "GL019383        CTAAGCCAGAGACTTAGTCAAAAGACAAAAGCCATCAGAGGAAAAA...\n",
       "GL019465        ATACTTGGCTCCCCACCCAGGAGATCGCAGCTTGAGGCATCACCCC...\n",
       "GL019525        TTTTAATGATGATGAAATCAGATCAGACCTATTGTCATGTTTAATC...\n",
       "dtype: object"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "numtless_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "af940afa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1               9\n",
       "11              1\n",
       "12              3\n",
       "13              2\n",
       "14              1\n",
       "15              5\n",
       "16              1\n",
       "17              3\n",
       "18              4\n",
       "19              1\n",
       "2               2\n",
       "20              2\n",
       "21              1\n",
       "3               1\n",
       "4               8\n",
       "7               3\n",
       "9               4\n",
       "AAGW02079688    3\n",
       "AAGW02081594    4\n",
       "AAGW02083714    1\n",
       "GL018700        1\n",
       "GL018705        1\n",
       "GL018706        1\n",
       "GL018741        5\n",
       "GL018744        1\n",
       "GL018748        3\n",
       "GL018786        1\n",
       "GL018817        4\n",
       "GL018826        8\n",
       "GL018862        2\n",
       "GL018925        2\n",
       "GL018992        1\n",
       "GL019010        3\n",
       "GL019047        3\n",
       "GL019056        1\n",
       "GL019086        1\n",
       "GL019149        2\n",
       "GL019235        1\n",
       "GL019271        1\n",
       "GL019308        5\n",
       "GL019311        2\n",
       "GL019326        6\n",
       "GL019383        2\n",
       "GL019465        5\n",
       "GL019525        1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "identifiers = pd.Series(numtless_sequences.index.values)\n",
    "numt_count = identifiers.apply(lambda g_id : list(numt_array['g_id']).count(g_id))\n",
    "numt_count.index = numtless_sequences.index.values\n",
    "numt_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "999ec7be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1               194841325\n",
       "11               87553407\n",
       "12              155354606\n",
       "13              143360353\n",
       "14              163896177\n",
       "15              109051050\n",
       "16               84478782\n",
       "17               85007110\n",
       "18               69798111\n",
       "19               57279035\n",
       "2               174331712\n",
       "20               33190240\n",
       "21               15577954\n",
       "3               155690876\n",
       "4                91385928\n",
       "7               173683288\n",
       "9               116249691\n",
       "AAGW02079688        72642\n",
       "AAGW02081594        16966\n",
       "AAGW02083714         5154\n",
       "GL018700         13460905\n",
       "GL018705          6123776\n",
       "GL018706          5787096\n",
       "GL018741          2346725\n",
       "GL018744          2220535\n",
       "GL018748          2145859\n",
       "GL018786          1165074\n",
       "GL018817           943909\n",
       "GL018826           917140\n",
       "GL018862           659056\n",
       "GL018925           585214\n",
       "GL018992           287663\n",
       "GL019010           276843\n",
       "GL019047           246181\n",
       "GL019056           198795\n",
       "GL019086           290499\n",
       "GL019149           135766\n",
       "GL019235           104204\n",
       "GL019271            83514\n",
       "GL019308            77650\n",
       "GL019311            79640\n",
       "GL019326            75171\n",
       "GL019383            58074\n",
       "GL019465            47584\n",
       "GL019525            42305\n",
       "dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lengths = numtless_sequences.apply(lambda sequence : len(sequence))\n",
    "lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "eb158d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function for sampling genome\n",
    "def sampling_genome(g_id):\n",
    "    global sum_ranges\n",
    "    number = numt_count[g_id]\n",
    "    sample_ranges = []\n",
    "    for i in range(0,number,1):\n",
    "        np.random.seed(i + 5)\n",
    "        length = lengths[g_id]\n",
    "        sequence = numtless_sequences[g_id]\n",
    "        range_max = length - 5000\n",
    "        range_start = np.random.randint(0, range_max)\n",
    "        range_end = range_start + 5000\n",
    "        sample_ranges.append(g_id + ':' + str(range_start) + '-' + str(range_end))\n",
    "    sum_ranges += sample_ranges\n",
    "    return sample_ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "022ebec7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1:148147043-148152043',\n",
       " '1:76708746-76713746',\n",
       " '1:59306159-59311159',\n",
       " '1:47736148-47741148',\n",
       " '1:44556670-44561670',\n",
       " '1:91571465-91576465',\n",
       " '1:24940223-24945223',\n",
       " '1:125253451-125258451',\n",
       " '1:118980946-118985946',\n",
       " '11:13929315-13934315',\n",
       " '12:148147043-148152043',\n",
       " '12:76708746-76713746',\n",
       " '12:59306159-59311159',\n",
       " '13:80864957-80869957',\n",
       " '13:76708746-76713746',\n",
       " '14:148147043-148152043',\n",
       " '15:13929315-13934315',\n",
       " '15:76708746-76713746',\n",
       " '15:59306159-59311159',\n",
       " '15:47736148-47741148',\n",
       " '15:44556670-44561670',\n",
       " '16:13929315-13934315',\n",
       " '17:13929315-13934315',\n",
       " '17:76708746-76713746',\n",
       " '17:59306159-59311159',\n",
       " '18:13929315-13934315',\n",
       " '18:42846921-42851921',\n",
       " '18:59306159-59311159',\n",
       " '18:47736148-47741148',\n",
       " '19:13929315-13934315',\n",
       " '2:148147043-148152043',\n",
       " '2:76708746-76713746',\n",
       " '20:13929315-13934315',\n",
       " '20:9599882-9604882',\n",
       " '21:13929315-13934315',\n",
       " '3:148147043-148152043',\n",
       " '4:13929315-13934315',\n",
       " '4:76708746-76713746',\n",
       " '4:59306159-59311159',\n",
       " '4:47736148-47741148',\n",
       " '4:44556670-44561670',\n",
       " '4:75209853-75214853',\n",
       " '4:24940223-24945223',\n",
       " '4:37459611-37464611',\n",
       " '7:148147043-148152043',\n",
       " '7:76708746-76713746',\n",
       " '7:59306159-59311159',\n",
       " '9:13929315-13934315',\n",
       " '9:76708746-76713746',\n",
       " '9:59306159-59311159',\n",
       " '9:47736148-47741148',\n",
       " 'AAGW02079688:35683-40683',\n",
       " 'AAGW02079688:31626-36626',\n",
       " 'AAGW02079688:61615-66615',\n",
       " 'AAGW02081594:2915-7915',\n",
       " 'AAGW02081594:2761-7761',\n",
       " 'AAGW02081594:9412-14412',\n",
       " 'AAGW02081594:4547-9547',\n",
       " 'AAGW02083714:99-5099',\n",
       " 'GL018700:2115790-2120790',\n",
       " 'GL018705:5540707-5545707',\n",
       " 'GL018706:5540707-5545707',\n",
       " 'GL018741:1346403-1351403',\n",
       " 'GL018741:1211274-1216274',\n",
       " 'GL018741:585903-590903',\n",
       " 'GL018741:1642947-1647947',\n",
       " 'GL018741:872796-877796',\n",
       " 'GL018744:1346403-1351403',\n",
       " 'GL018748:1346403-1351403',\n",
       " 'GL018748:1211274-1216274',\n",
       " 'GL018748:585903-590903',\n",
       " 'GL018786:18638-23638',\n",
       " 'GL018817:297827-302827',\n",
       " 'GL018817:162698-167698',\n",
       " 'GL018817:585903-590903',\n",
       " 'GL018817:594371-599371',\n",
       " 'GL018826:297827-302827',\n",
       " 'GL018826:162698-167698',\n",
       " 'GL018826:585903-590903',\n",
       " 'GL018826:594371-599371',\n",
       " 'GL018826:516478-521478',\n",
       " 'GL018826:345353-350353',\n",
       " 'GL018826:403353-408353',\n",
       " 'GL018826:472907-477907',\n",
       " 'GL018862:297827-302827',\n",
       " 'GL018862:162698-167698',\n",
       " 'GL018925:297827-302827',\n",
       " 'GL018925:162698-167698',\n",
       " 'GL018992:18638-23638',\n",
       " 'GL019010:18638-23638',\n",
       " 'GL019010:162698-167698',\n",
       " 'GL019010:61615-66615',\n",
       " 'GL019047:35683-40683',\n",
       " 'GL019047:162698-167698',\n",
       " 'GL019047:61615-66615',\n",
       " 'GL019056:35683-40683',\n",
       " 'GL019086:18638-23638',\n",
       " 'GL019149:35683-40683',\n",
       " 'GL019149:31626-36626',\n",
       " 'GL019235:35683-40683',\n",
       " 'GL019271:35683-40683',\n",
       " 'GL019308:35683-40683',\n",
       " 'GL019308:31626-36626',\n",
       " 'GL019308:61615-66615',\n",
       " 'GL019308:70083-75083',\n",
       " 'GL019308:55934-60934',\n",
       " 'GL019311:35683-40683',\n",
       " 'GL019311:31626-36626',\n",
       " 'GL019326:35683-40683',\n",
       " 'GL019326:31626-36626',\n",
       " 'GL019326:61615-66615',\n",
       " 'GL019326:70083-75083',\n",
       " 'GL019326:55934-60934',\n",
       " 'GL019326:50496-55496',\n",
       " 'GL019383:35683-40683',\n",
       " 'GL019383:31626-36626',\n",
       " 'GL019465:35683-40683',\n",
       " 'GL019465:31626-36626',\n",
       " 'GL019465:10742-15742',\n",
       " 'GL019465:4547-9547',\n",
       " 'GL019465:20828-25828',\n",
       " 'GL019525:35683-40683']"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#get the sample ranges\n",
    "sum_ranges = []\n",
    "identifiers.index = identifiers\n",
    "samples = identifiers.apply(sampling_genome)\n",
    "sum_ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "e4d520d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#writing output file\n",
    "with open(home_dir + 'repeat_samples.bed', 'w') as outfile:\n",
    "    pd.Series(sum_ranges).apply(lambda sample_range : outfile.write(sample_range + '\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "ca7ed94d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genoName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>275</td>\n",
       "      <td>262</td>\n",
       "      <td>119</td>\n",
       "      <td>0</td>\n",
       "      <td>114727071</td>\n",
       "      <td>114727155</td>\n",
       "      <td>-80123602</td>\n",
       "      <td>+</td>\n",
       "      <td>MLT1A</td>\n",
       "      <td>LTR</td>\n",
       "      <td>ERVL-MaLR</td>\n",
       "      <td>25</td>\n",
       "      <td>118</td>\n",
       "      <td>-256</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>1141</td>\n",
       "      <td>268</td>\n",
       "      <td>45</td>\n",
       "      <td>3</td>\n",
       "      <td>114727222</td>\n",
       "      <td>114727559</td>\n",
       "      <td>-80123198</td>\n",
       "      <td>-</td>\n",
       "      <td>MLT-int</td>\n",
       "      <td>LTR</td>\n",
       "      <td>ERVL-MaLR</td>\n",
       "      <td>-12</td>\n",
       "      <td>1723</td>\n",
       "      <td>1373</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>35</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>114728330</td>\n",
       "      <td>114728393</td>\n",
       "      <td>-80122364</td>\n",
       "      <td>+</td>\n",
       "      <td>AT_rich</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>1</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>114728439</td>\n",
       "      <td>114728461</td>\n",
       "      <td>-80122296</td>\n",
       "      <td>+</td>\n",
       "      <td>AT_rich</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>Low_complexity</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>957</td>\n",
       "      <td>226</td>\n",
       "      <td>61</td>\n",
       "      <td>19</td>\n",
       "      <td>114728509</td>\n",
       "      <td>114728771</td>\n",
       "      <td>-80121986</td>\n",
       "      <td>-</td>\n",
       "      <td>CSINE1</td>\n",
       "      <td>SINE</td>\n",
       "      <td>SINE</td>\n",
       "      <td>-33</td>\n",
       "      <td>307</td>\n",
       "      <td>35</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          #bin  swScore  milliDiv  milliDel  milliIns  genoStart    genoEnd  \\\n",
       "genoName                                                                      \n",
       "chr1      1460      275       262       119         0  114727071  114727155   \n",
       "chr1      1460     1141       268        45         3  114727222  114727559   \n",
       "chr1      1460       35        63         0         0  114728330  114728393   \n",
       "chr1      1460       22         0         0         0  114728439  114728461   \n",
       "chr1      1460      957       226        61        19  114728509  114728771   \n",
       "\n",
       "          genoLeft strand  repName        repClass       repFamily  repStart  \\\n",
       "genoName                                                                       \n",
       "chr1     -80123602      +    MLT1A             LTR       ERVL-MaLR        25   \n",
       "chr1     -80123198      -  MLT-int             LTR       ERVL-MaLR       -12   \n",
       "chr1     -80122364      +  AT_rich  Low_complexity  Low_complexity         1   \n",
       "chr1     -80122296      +  AT_rich  Low_complexity  Low_complexity         1   \n",
       "chr1     -80121986      -   CSINE1            SINE            SINE       -33   \n",
       "\n",
       "          repEnd  repLeft  id  \n",
       "genoName                       \n",
       "chr1         118     -256   1  \n",
       "chr1        1723     1373   1  \n",
       "chr1          63        0   1  \n",
       "chr1          22        0   1  \n",
       "chr1         307       35   1  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read upstream repeats\n",
    "upstream_repeats = pd.read_csv(home_dir + r'repeats/upstream_flanking_repeats', sep = '\\t')\n",
    "upstream_repeats = upstream_repeats.set_index('genoName')\n",
    "upstream_repeats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "061fb366",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genoName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>416</td>\n",
       "      <td>122</td>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>114736112</td>\n",
       "      <td>114736195</td>\n",
       "      <td>-80114562</td>\n",
       "      <td>+</td>\n",
       "      <td>(TA)n</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>Simple_repeat</td>\n",
       "      <td>1</td>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>1560</td>\n",
       "      <td>213</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "      <td>114736195</td>\n",
       "      <td>114736552</td>\n",
       "      <td>-80114205</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M2</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>4808</td>\n",
       "      <td>5179</td>\n",
       "      <td>-964</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>283</td>\n",
       "      <td>283</td>\n",
       "      <td>35</td>\n",
       "      <td>17</td>\n",
       "      <td>114736561</td>\n",
       "      <td>114736676</td>\n",
       "      <td>-80114081</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M5</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>5526</td>\n",
       "      <td>5642</td>\n",
       "      <td>-600</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>515</td>\n",
       "      <td>280</td>\n",
       "      <td>5</td>\n",
       "      <td>107</td>\n",
       "      <td>114736709</td>\n",
       "      <td>114736801</td>\n",
       "      <td>-80113956</td>\n",
       "      <td>+</td>\n",
       "      <td>CSINE2</td>\n",
       "      <td>SINE</td>\n",
       "      <td>SINE</td>\n",
       "      <td>5</td>\n",
       "      <td>80</td>\n",
       "      <td>-264</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>1698</td>\n",
       "      <td>114</td>\n",
       "      <td>99</td>\n",
       "      <td>19</td>\n",
       "      <td>114736801</td>\n",
       "      <td>114737115</td>\n",
       "      <td>-80113642</td>\n",
       "      <td>-</td>\n",
       "      <td>CSINE3A</td>\n",
       "      <td>SINE</td>\n",
       "      <td>SINE</td>\n",
       "      <td>-11</td>\n",
       "      <td>339</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          #bin  swScore  milliDiv  milliDel  milliIns  genoStart    genoEnd  \\\n",
       "genoName                                                                      \n",
       "chr1      1460      416       122         0        12  114736112  114736195   \n",
       "chr1      1460     1560       213        42         0  114736195  114736552   \n",
       "chr1      1460      283       283        35        17  114736561  114736676   \n",
       "chr1      1460      515       280         5       107  114736709  114736801   \n",
       "chr1      1460     1698       114        99        19  114736801  114737115   \n",
       "\n",
       "          genoLeft strand  repName       repClass      repFamily  repStart  \\\n",
       "genoName                                                                     \n",
       "chr1     -80114562      +    (TA)n  Simple_repeat  Simple_repeat         1   \n",
       "chr1     -80114205      +     L1M2           LINE             L1      4808   \n",
       "chr1     -80114081      +     L1M5           LINE             L1      5526   \n",
       "chr1     -80113956      +   CSINE2           SINE           SINE         5   \n",
       "chr1     -80113642      -  CSINE3A           SINE           SINE       -11   \n",
       "\n",
       "          repEnd  repLeft  id  \n",
       "genoName                       \n",
       "chr1          82        0   1  \n",
       "chr1        5179     -964   1  \n",
       "chr1        5642     -600   1  \n",
       "chr1          80     -264   1  \n",
       "chr1         339        1   1  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read downstream repeats\n",
    "downstream_repeats = pd.read_csv(home_dir + r'repeats/downstream_flanking_repeats', sep = '\\t')\n",
    "downstream_repeats = downstream_repeats.set_index('genoName')\n",
    "downstream_repeats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6e6f3baa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genoName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1170</td>\n",
       "      <td>1094</td>\n",
       "      <td>181</td>\n",
       "      <td>8</td>\n",
       "      <td>96</td>\n",
       "      <td>76710292</td>\n",
       "      <td>76710543</td>\n",
       "      <td>-118140214</td>\n",
       "      <td>+</td>\n",
       "      <td>CSINE2</td>\n",
       "      <td>SINE</td>\n",
       "      <td>SINE</td>\n",
       "      <td>116</td>\n",
       "      <td>344</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1170</td>\n",
       "      <td>666</td>\n",
       "      <td>297</td>\n",
       "      <td>105</td>\n",
       "      <td>47</td>\n",
       "      <td>76710539</td>\n",
       "      <td>76711053</td>\n",
       "      <td>-118139704</td>\n",
       "      <td>+</td>\n",
       "      <td>MLT1E2</td>\n",
       "      <td>LTR</td>\n",
       "      <td>ERVL-MaLR</td>\n",
       "      <td>73</td>\n",
       "      <td>627</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1170</td>\n",
       "      <td>6358</td>\n",
       "      <td>220</td>\n",
       "      <td>58</td>\n",
       "      <td>24</td>\n",
       "      <td>76712233</td>\n",
       "      <td>76712345</td>\n",
       "      <td>-118138412</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M2</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>2379</td>\n",
       "      <td>2489</td>\n",
       "      <td>-3654</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1170</td>\n",
       "      <td>2513</td>\n",
       "      <td>57</td>\n",
       "      <td>48</td>\n",
       "      <td>6</td>\n",
       "      <td>76712345</td>\n",
       "      <td>76712679</td>\n",
       "      <td>-118138078</td>\n",
       "      <td>+</td>\n",
       "      <td>CSINE3A</td>\n",
       "      <td>SINE</td>\n",
       "      <td>SINE</td>\n",
       "      <td>1</td>\n",
       "      <td>348</td>\n",
       "      <td>-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1170</td>\n",
       "      <td>6358</td>\n",
       "      <td>220</td>\n",
       "      <td>58</td>\n",
       "      <td>24</td>\n",
       "      <td>76712679</td>\n",
       "      <td>76713312</td>\n",
       "      <td>-118137445</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M2</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>2490</td>\n",
       "      <td>3147</td>\n",
       "      <td>-2996</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          #bin  swScore  milliDiv  milliDel  milliIns  genoStart   genoEnd  \\\n",
       "genoName                                                                     \n",
       "chr1      1170     1094       181         8        96   76710292  76710543   \n",
       "chr1      1170      666       297       105        47   76710539  76711053   \n",
       "chr1      1170     6358       220        58        24   76712233  76712345   \n",
       "chr1      1170     2513        57        48         6   76712345  76712679   \n",
       "chr1      1170     6358       220        58        24   76712679  76713312   \n",
       "\n",
       "           genoLeft strand  repName repClass  repFamily  repStart  repEnd  \\\n",
       "genoName                                                                    \n",
       "chr1     -118140214      +   CSINE2     SINE       SINE       116     344   \n",
       "chr1     -118139704      +   MLT1E2      LTR  ERVL-MaLR        73     627   \n",
       "chr1     -118138412      +     L1M2     LINE         L1      2379    2489   \n",
       "chr1     -118138078      +  CSINE3A     SINE       SINE         1     348   \n",
       "chr1     -118137445      +     L1M2     LINE         L1      2490    3147   \n",
       "\n",
       "          repLeft  id  \n",
       "genoName               \n",
       "chr1            0   1  \n",
       "chr1            0   1  \n",
       "chr1        -3654   1  \n",
       "chr1           -2   1  \n",
       "chr1        -2996   1  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read sample repeats\n",
    "sample_repeats = pd.read_csv(home_dir + r'repeats/sample_repeats', sep = '\\t')\n",
    "sample_repeats = sample_repeats.set_index('genoName')\n",
    "sample_repeats.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "4da4ac66",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get common repetitions in all three dataset (upstream, downstream, genome)\n",
    "common_reps = pd.Series(list(set(upstream_repeats['repName'].unique()) &\n",
    "                             set(downstream_repeats['repName'].unique()) &\n",
    "                             set(sample_repeats['repName'].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "8b6849d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_repeatnumber(chromosome, repname):\n",
    "    up_sub = upstream_repeats.loc[chromosome]\n",
    "    down_sub = downstream_repeats.loc[chromosome]\n",
    "    sample_sub = sample_repeats.loc[chromosome]\n",
    "    return [list(up_sub['repName']).count(repname),\n",
    "           list(down_sub['repName']).count(repname),\n",
    "           list(sample_sub['repName']).count(repname)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "027e5643",
   "metadata": {},
   "outputs": [],
   "source": [
    "#empty series for the repetitions\n",
    "repnums = pd.Series(index = common_reps, dtype = str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "3694d57a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for repname in common_reps:\n",
    "    chromosomes = pd.Series(np.unique(sample_repeats.index.values))\n",
    "    samples = chromosomes.apply(get_repeatnumber, args = (repname,))\n",
    "    up_reps = []\n",
    "    samples.apply(lambda sample : up_reps.append(sample[0]))\n",
    "    down_reps = []\n",
    "    samples.apply(lambda sample : down_reps.append(sample[1]))\n",
    "    sample_reps = []\n",
    "    samples.apply(lambda sample : sample_reps.append(sample[2]))\n",
    "    repnums[repname] = [up_reps, down_reps, sample_reps]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "0784351a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#function for the statistical analysis of repetitions\n",
    "#upstream_repetitions = nested_list[0]\n",
    "#downstream_repetitions = nested_list[1]\n",
    "#sample_repetitions = nested_list[2]\n",
    "def statistical_analysis(nested_list, which):\n",
    "    sample1 = nested_list[which[0]]\n",
    "    sample2 = nested_list[which[1]]\n",
    "    norm1 = anderson(sample1)\n",
    "    norm2 = anderson(sample2)\n",
    "    stat1 = norm1[0]\n",
    "    stat2 = norm2[0]\n",
    "    critical_value1 = norm1[1][2]\n",
    "    critical_value2 = norm2[1][2]\n",
    "    if (stat1 > critical_value1) or (stat2 > critical_value2):\n",
    "        try:\n",
    "            significance = wilcoxon(sample1, sample2)[1]\n",
    "            return significance\n",
    "        except ValueError:\n",
    "            pass\n",
    "    else:\n",
    "        try:\n",
    "            significance = ttest_ind(sample1, sample2)[1]\n",
    "            return significance\n",
    "        except ValueError:\n",
    "            pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "2907f097",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MIRb     0.002682\n",
       "(TG)n    0.039008\n",
       "dtype: float64"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#statistics of upstream and downstream repetitions\n",
    "ud_stat = repnums.apply(statistical_analysis, args = ([0,1],))\n",
    "ud_stat[ud_stat < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "58644a90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genoName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chrUn0634</th>\n",
       "      <td>585</td>\n",
       "      <td>186</td>\n",
       "      <td>289</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>58945</td>\n",
       "      <td>58990</td>\n",
       "      <td>-29238</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-23</td>\n",
       "      <td>245</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn1218</th>\n",
       "      <td>585</td>\n",
       "      <td>186</td>\n",
       "      <td>289</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>10976</td>\n",
       "      <td>11021</td>\n",
       "      <td>-13555</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-23</td>\n",
       "      <td>245</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0658</th>\n",
       "      <td>585</td>\n",
       "      <td>186</td>\n",
       "      <td>289</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>54933</td>\n",
       "      <td>54978</td>\n",
       "      <td>-26744</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-23</td>\n",
       "      <td>245</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0634</th>\n",
       "      <td>585</td>\n",
       "      <td>186</td>\n",
       "      <td>289</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>58945</td>\n",
       "      <td>58990</td>\n",
       "      <td>-29238</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-23</td>\n",
       "      <td>245</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0658</th>\n",
       "      <td>585</td>\n",
       "      <td>186</td>\n",
       "      <td>289</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>54933</td>\n",
       "      <td>54978</td>\n",
       "      <td>-26744</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-23</td>\n",
       "      <td>245</td>\n",
       "      <td>200</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9</th>\n",
       "      <td>912</td>\n",
       "      <td>379</td>\n",
       "      <td>293</td>\n",
       "      <td>68</td>\n",
       "      <td>8</td>\n",
       "      <td>42950966</td>\n",
       "      <td>42951170</td>\n",
       "      <td>-73300737</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>23</td>\n",
       "      <td>251</td>\n",
       "      <td>-17</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0088</th>\n",
       "      <td>586</td>\n",
       "      <td>255</td>\n",
       "      <td>364</td>\n",
       "      <td>53</td>\n",
       "      <td>8</td>\n",
       "      <td>214523</td>\n",
       "      <td>214656</td>\n",
       "      <td>-950629</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>40</td>\n",
       "      <td>178</td>\n",
       "      <td>-90</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13</th>\n",
       "      <td>700</td>\n",
       "      <td>229</td>\n",
       "      <td>330</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>15103333</td>\n",
       "      <td>15103433</td>\n",
       "      <td>-128257399</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-156</td>\n",
       "      <td>112</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7</th>\n",
       "      <td>1550</td>\n",
       "      <td>284</td>\n",
       "      <td>320</td>\n",
       "      <td>96</td>\n",
       "      <td>10</td>\n",
       "      <td>126584170</td>\n",
       "      <td>126584274</td>\n",
       "      <td>-47100185</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-40</td>\n",
       "      <td>228</td>\n",
       "      <td>116</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr4</th>\n",
       "      <td>721</td>\n",
       "      <td>286</td>\n",
       "      <td>328</td>\n",
       "      <td>76</td>\n",
       "      <td>0</td>\n",
       "      <td>17871713</td>\n",
       "      <td>17871832</td>\n",
       "      <td>-73522268</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>14</td>\n",
       "      <td>141</td>\n",
       "      <td>-127</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr18</th>\n",
       "      <td>1011</td>\n",
       "      <td>189</td>\n",
       "      <td>330</td>\n",
       "      <td>68</td>\n",
       "      <td>25</td>\n",
       "      <td>55872812</td>\n",
       "      <td>55872930</td>\n",
       "      <td>-13927806</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>88</td>\n",
       "      <td>210</td>\n",
       "      <td>-58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20</th>\n",
       "      <td>642</td>\n",
       "      <td>265</td>\n",
       "      <td>271</td>\n",
       "      <td>70</td>\n",
       "      <td>49</td>\n",
       "      <td>7601908</td>\n",
       "      <td>7602047</td>\n",
       "      <td>-25589285</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>59</td>\n",
       "      <td>216</td>\n",
       "      <td>-52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr20</th>\n",
       "      <td>642</td>\n",
       "      <td>265</td>\n",
       "      <td>271</td>\n",
       "      <td>70</td>\n",
       "      <td>49</td>\n",
       "      <td>7601908</td>\n",
       "      <td>7602047</td>\n",
       "      <td>-25589285</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>59</td>\n",
       "      <td>216</td>\n",
       "      <td>-52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0128</th>\n",
       "      <td>588</td>\n",
       "      <td>289</td>\n",
       "      <td>232</td>\n",
       "      <td>95</td>\n",
       "      <td>57</td>\n",
       "      <td>410291</td>\n",
       "      <td>410396</td>\n",
       "      <td>-511819</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0128</th>\n",
       "      <td>586</td>\n",
       "      <td>250</td>\n",
       "      <td>280</td>\n",
       "      <td>86</td>\n",
       "      <td>48</td>\n",
       "      <td>138632</td>\n",
       "      <td>138737</td>\n",
       "      <td>-783478</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0128</th>\n",
       "      <td>586</td>\n",
       "      <td>250</td>\n",
       "      <td>280</td>\n",
       "      <td>86</td>\n",
       "      <td>48</td>\n",
       "      <td>138632</td>\n",
       "      <td>138737</td>\n",
       "      <td>-783478</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0351</th>\n",
       "      <td>586</td>\n",
       "      <td>210</td>\n",
       "      <td>290</td>\n",
       "      <td>86</td>\n",
       "      <td>48</td>\n",
       "      <td>204642</td>\n",
       "      <td>204747</td>\n",
       "      <td>-42681</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11</th>\n",
       "      <td>829</td>\n",
       "      <td>190</td>\n",
       "      <td>226</td>\n",
       "      <td>73</td>\n",
       "      <td>88</td>\n",
       "      <td>32101710</td>\n",
       "      <td>32101778</td>\n",
       "      <td>-55452436</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>39</td>\n",
       "      <td>105</td>\n",
       "      <td>-163</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0128</th>\n",
       "      <td>588</td>\n",
       "      <td>289</td>\n",
       "      <td>232</td>\n",
       "      <td>95</td>\n",
       "      <td>57</td>\n",
       "      <td>410291</td>\n",
       "      <td>410396</td>\n",
       "      <td>-511819</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0128</th>\n",
       "      <td>586</td>\n",
       "      <td>250</td>\n",
       "      <td>280</td>\n",
       "      <td>86</td>\n",
       "      <td>48</td>\n",
       "      <td>138632</td>\n",
       "      <td>138737</td>\n",
       "      <td>-783478</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2</th>\n",
       "      <td>875</td>\n",
       "      <td>489</td>\n",
       "      <td>305</td>\n",
       "      <td>28</td>\n",
       "      <td>61</td>\n",
       "      <td>38018247</td>\n",
       "      <td>38018460</td>\n",
       "      <td>-136313852</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>20</td>\n",
       "      <td>225</td>\n",
       "      <td>-43</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3</th>\n",
       "      <td>756</td>\n",
       "      <td>395</td>\n",
       "      <td>278</td>\n",
       "      <td>34</td>\n",
       "      <td>17</td>\n",
       "      <td>22496655</td>\n",
       "      <td>22496772</td>\n",
       "      <td>-133194333</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>40</td>\n",
       "      <td>158</td>\n",
       "      <td>-110</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3</th>\n",
       "      <td>756</td>\n",
       "      <td>534</td>\n",
       "      <td>209</td>\n",
       "      <td>77</td>\n",
       "      <td>63</td>\n",
       "      <td>22500023</td>\n",
       "      <td>22500166</td>\n",
       "      <td>-133190939</td>\n",
       "      <td>-</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>-45</td>\n",
       "      <td>223</td>\n",
       "      <td>79</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3</th>\n",
       "      <td>756</td>\n",
       "      <td>197</td>\n",
       "      <td>402</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>22500351</td>\n",
       "      <td>22500463</td>\n",
       "      <td>-133190642</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>6</td>\n",
       "      <td>123</td>\n",
       "      <td>-145</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15</th>\n",
       "      <td>1347</td>\n",
       "      <td>214</td>\n",
       "      <td>336</td>\n",
       "      <td>45</td>\n",
       "      <td>27</td>\n",
       "      <td>99933516</td>\n",
       "      <td>99933626</td>\n",
       "      <td>-9120426</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>51</td>\n",
       "      <td>162</td>\n",
       "      <td>-106</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0351</th>\n",
       "      <td>586</td>\n",
       "      <td>210</td>\n",
       "      <td>290</td>\n",
       "      <td>86</td>\n",
       "      <td>48</td>\n",
       "      <td>204642</td>\n",
       "      <td>204747</td>\n",
       "      <td>-42681</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>8</td>\n",
       "      <td>116</td>\n",
       "      <td>-152</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0046</th>\n",
       "      <td>590</td>\n",
       "      <td>230</td>\n",
       "      <td>329</td>\n",
       "      <td>122</td>\n",
       "      <td>14</td>\n",
       "      <td>669931</td>\n",
       "      <td>670079</td>\n",
       "      <td>-1550776</td>\n",
       "      <td>+</td>\n",
       "      <td>MIRb</td>\n",
       "      <td>SINE</td>\n",
       "      <td>MIR</td>\n",
       "      <td>18</td>\n",
       "      <td>181</td>\n",
       "      <td>-87</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           #bin  swScore  milliDiv  milliDel  milliIns  genoStart    genoEnd  \\\n",
       "genoName                                                                       \n",
       "chrUn0634   585      186       289        22         0      58945      58990   \n",
       "chrUn1218   585      186       289        22         0      10976      11021   \n",
       "chrUn0658   585      186       289        22         0      54933      54978   \n",
       "chrUn0634   585      186       289        22         0      58945      58990   \n",
       "chrUn0658   585      186       289        22         0      54933      54978   \n",
       "chr9        912      379       293        68         8   42950966   42951170   \n",
       "chrUn0088   586      255       364        53         8     214523     214656   \n",
       "chr13       700      229       330        40         0   15103333   15103433   \n",
       "chr7       1550      284       320        96        10  126584170  126584274   \n",
       "chr4        721      286       328        76         0   17871713   17871832   \n",
       "chr18      1011      189       330        68        25   55872812   55872930   \n",
       "chr20       642      265       271        70        49    7601908    7602047   \n",
       "chr20       642      265       271        70        49    7601908    7602047   \n",
       "chrUn0128   588      289       232        95        57     410291     410396   \n",
       "chrUn0128   586      250       280        86        48     138632     138737   \n",
       "chrUn0128   586      250       280        86        48     138632     138737   \n",
       "chrUn0351   586      210       290        86        48     204642     204747   \n",
       "chr11       829      190       226        73        88   32101710   32101778   \n",
       "chrUn0128   588      289       232        95        57     410291     410396   \n",
       "chrUn0128   586      250       280        86        48     138632     138737   \n",
       "chr2        875      489       305        28        61   38018247   38018460   \n",
       "chr3        756      395       278        34        17   22496655   22496772   \n",
       "chr3        756      534       209        77        63   22500023   22500166   \n",
       "chr3        756      197       402        54         0   22500351   22500463   \n",
       "chr15      1347      214       336        45        27   99933516   99933626   \n",
       "chrUn0351   586      210       290        86        48     204642     204747   \n",
       "chrUn0046   590      230       329       122        14     669931     670079   \n",
       "\n",
       "            genoLeft strand repName repClass repFamily  repStart  repEnd  \\\n",
       "genoName                                                                   \n",
       "chrUn0634     -29238      -    MIRb     SINE       MIR       -23     245   \n",
       "chrUn1218     -13555      -    MIRb     SINE       MIR       -23     245   \n",
       "chrUn0658     -26744      -    MIRb     SINE       MIR       -23     245   \n",
       "chrUn0634     -29238      -    MIRb     SINE       MIR       -23     245   \n",
       "chrUn0658     -26744      -    MIRb     SINE       MIR       -23     245   \n",
       "chr9       -73300737      +    MIRb     SINE       MIR        23     251   \n",
       "chrUn0088    -950629      +    MIRb     SINE       MIR        40     178   \n",
       "chr13     -128257399      -    MIRb     SINE       MIR      -156     112   \n",
       "chr7       -47100185      -    MIRb     SINE       MIR       -40     228   \n",
       "chr4       -73522268      +    MIRb     SINE       MIR        14     141   \n",
       "chr18      -13927806      +    MIRb     SINE       MIR        88     210   \n",
       "chr20      -25589285      +    MIRb     SINE       MIR        59     216   \n",
       "chr20      -25589285      +    MIRb     SINE       MIR        59     216   \n",
       "chrUn0128    -511819      +    MIRb     SINE       MIR         8     116   \n",
       "chrUn0128    -783478      +    MIRb     SINE       MIR         8     116   \n",
       "chrUn0128    -783478      +    MIRb     SINE       MIR         8     116   \n",
       "chrUn0351     -42681      +    MIRb     SINE       MIR         8     116   \n",
       "chr11      -55452436      +    MIRb     SINE       MIR        39     105   \n",
       "chrUn0128    -511819      +    MIRb     SINE       MIR         8     116   \n",
       "chrUn0128    -783478      +    MIRb     SINE       MIR         8     116   \n",
       "chr2      -136313852      +    MIRb     SINE       MIR        20     225   \n",
       "chr3      -133194333      +    MIRb     SINE       MIR        40     158   \n",
       "chr3      -133190939      -    MIRb     SINE       MIR       -45     223   \n",
       "chr3      -133190642      +    MIRb     SINE       MIR         6     123   \n",
       "chr15       -9120426      +    MIRb     SINE       MIR        51     162   \n",
       "chrUn0351     -42681      +    MIRb     SINE       MIR         8     116   \n",
       "chrUn0046   -1550776      +    MIRb     SINE       MIR        18     181   \n",
       "\n",
       "           repLeft  id  \n",
       "genoName                \n",
       "chrUn0634      200   4  \n",
       "chrUn1218      200   4  \n",
       "chrUn0658      200   4  \n",
       "chrUn0634      200   4  \n",
       "chrUn0658      200   4  \n",
       "chr9           -17   3  \n",
       "chrUn0088      -90   3  \n",
       "chr13            9   7  \n",
       "chr7           116   2  \n",
       "chr4          -127   2  \n",
       "chr18          -58   1  \n",
       "chr20          -52   2  \n",
       "chr20          -52   2  \n",
       "chrUn0128     -152   3  \n",
       "chrUn0128     -152   3  \n",
       "chrUn0128     -152   3  \n",
       "chrUn0351     -152   3  \n",
       "chr11         -163   4  \n",
       "chrUn0128     -152   3  \n",
       "chrUn0128     -152   3  \n",
       "chr2           -43   1  \n",
       "chr3          -110   2  \n",
       "chr3            79   2  \n",
       "chr3          -145   2  \n",
       "chr15         -106   1  \n",
       "chrUn0351     -152   3  \n",
       "chrUn0046      -87   3  "
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##just to check what is the class and family of 'MIRb'\n",
    "upstream_repeats.loc[upstream_repeats['repName'] == 'MIRb']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "264d07d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pytho\\Anaconda3\\lib\\site-packages\\scipy\\stats\\morestats.py:2981: UserWarning: Sample size too small for normal approximation.\n",
      "  warnings.warn(\"Sample size too small for normal approximation.\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "MIR3    0.006656\n",
       "MIR     0.007536\n",
       "dtype: float64"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#statistics of upstream and sample repetitions\n",
    "us_stat = repnums.apply(statistical_analysis, args = ([0,2],))\n",
    "us_stat[us_stat < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "6158a630",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\pytho\\Anaconda3\\lib\\site-packages\\scipy\\stats\\morestats.py:2981: UserWarning: Sample size too small for normal approximation.\n",
      "  warnings.warn(\"Sample size too small for normal approximation.\")\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "L1M5     0.045737\n",
       "MIR3     0.039360\n",
       "(TG)n    0.048140\n",
       "dtype: float64"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#statistics of downstream and sample repetitions\n",
    "ds_stat = repnums.apply(statistical_analysis, args = ([1,2],))\n",
    "ds_stat[ds_stat < 0.05]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "552e1440",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#bin</th>\n",
       "      <th>swScore</th>\n",
       "      <th>milliDiv</th>\n",
       "      <th>milliDel</th>\n",
       "      <th>milliIns</th>\n",
       "      <th>genoStart</th>\n",
       "      <th>genoEnd</th>\n",
       "      <th>genoLeft</th>\n",
       "      <th>strand</th>\n",
       "      <th>repName</th>\n",
       "      <th>repClass</th>\n",
       "      <th>repFamily</th>\n",
       "      <th>repStart</th>\n",
       "      <th>repEnd</th>\n",
       "      <th>repLeft</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genoName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1</th>\n",
       "      <td>1460</td>\n",
       "      <td>283</td>\n",
       "      <td>283</td>\n",
       "      <td>35</td>\n",
       "      <td>17</td>\n",
       "      <td>114736561</td>\n",
       "      <td>114736676</td>\n",
       "      <td>-80114081</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M5</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>5526</td>\n",
       "      <td>5642</td>\n",
       "      <td>-600</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrUn0634</th>\n",
       "      <td>585</td>\n",
       "      <td>246</td>\n",
       "      <td>192</td>\n",
       "      <td>122</td>\n",
       "      <td>49</td>\n",
       "      <td>76663</td>\n",
       "      <td>76745</td>\n",
       "      <td>-11483</td>\n",
       "      <td>+</td>\n",
       "      <td>L1M5</td>\n",
       "      <td>LINE</td>\n",
       "      <td>L1</td>\n",
       "      <td>5734</td>\n",
       "      <td>5821</td>\n",
       "      <td>-343</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           #bin  swScore  milliDiv  milliDel  milliIns  genoStart    genoEnd  \\\n",
       "genoName                                                                       \n",
       "chr1       1460      283       283        35        17  114736561  114736676   \n",
       "chrUn0634   585      246       192       122        49      76663      76745   \n",
       "\n",
       "           genoLeft strand repName repClass repFamily  repStart  repEnd  \\\n",
       "genoName                                                                  \n",
       "chr1      -80114081      +    L1M5     LINE        L1      5526    5642   \n",
       "chrUn0634    -11483      +    L1M5     LINE        L1      5734    5821   \n",
       "\n",
       "           repLeft  id  \n",
       "genoName                \n",
       "chr1          -600   1  \n",
       "chrUn0634     -343   4  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#just to check what is the class and family of 'L1M5'\n",
    "downstream_repeats.loc[downstream_repeats['repName'] == 'L1M5']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7c8bdc31",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 2, 0, 3, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 5, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 2, 0, 0, 2, 0, 0, 0, 1, 0]\n",
      "[0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n"
     ]
    }
   ],
   "source": [
    "print(repnums['MIRb'][0])\n",
    "print(repnums['MIRb'][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75acbfe5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
