{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Numt regions are specific genomic or mitochondrial regions where multiple numts are derived from. Practically it means the number of numt contributing nucleotides of the current region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing the required modules\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>eg2_value</th>\n",
       "      <th>e_value</th>\n",
       "      <th>g_id</th>\n",
       "      <th>g_start</th>\n",
       "      <th>mt_start</th>\n",
       "      <th>g_length</th>\n",
       "      <th>mt_length</th>\n",
       "      <th>g_strand</th>\n",
       "      <th>mt_strand</th>\n",
       "      <th>g_size</th>\n",
       "      <th>g_sequence</th>\n",
       "      <th>mt_sequence</th>\n",
       "      <th>g_up_flanking</th>\n",
       "      <th>g_down_flanking</th>\n",
       "      <th>mt_up_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "      <th>mt_down_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>17245</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>MT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>17245</td>\n",
       "      <td>17245</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>17245</td>\n",
       "      <td>GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...</td>\n",
       "      <td>GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TACAAATTGTATAATATTTGGACTAAACCCCCTTTCCCACCCCAAG...</td>\n",
       "      <td>GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...</td>\n",
       "      <td>ENSOCUG00000029081</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>114731432</td>\n",
       "      <td>6976</td>\n",
       "      <td>4680</td>\n",
       "      <td>4681</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>194850757</td>\n",
       "      <td>taGTAAAATTATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...</td>\n",
       "      <td>TAGTAAAACCATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...</td>\n",
       "      <td>ACAAAAGACTCTGAATAGCCAAAGCGATCCTGAACAAGAAAAATCA...</td>\n",
       "      <td>TATATATATTTATATATATATGTATATATGTACATATATATAACTA...</td>\n",
       "      <td>CCTTCGCCTCAAAACGAGAAGTAGAGACTATTGAACTAACAACCAC...</td>\n",
       "      <td>AACTGCTAATTCATGCTCCCATGTCTAACCCCATGGCTTTCTCAAC...</td>\n",
       "      <td>ENSOCUG00000006104</td>\n",
       "      <td>cilia and flagella associated protein 300 [Sou...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3120</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GL019308</td>\n",
       "      <td>67160</td>\n",
       "      <td>883</td>\n",
       "      <td>4892</td>\n",
       "      <td>4945</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>88228</td>\n",
       "      <td>CAAGGCCATGAAGCATGCACACACTGCCCATCGCTCTCCTCAAACA...</td>\n",
       "      <td>CAAGGCCATGAAGCACGCACACACCGCCCGTCACCCTCCTCAAGTG...</td>\n",
       "      <td>GATAAACCCCGATAAACTTTACCACTCTTTGCCAACTCAGCCTCTA...</td>\n",
       "      <td>ATGAAGGGGCAAAGAGTAATTTTACAGTCGAGAGGGCTAGCAGTCA...</td>\n",
       "      <td>TACCGCCATCTTCAGCGAACCCTAAAAAGGAGCAAAAGTAAGCTCA...</td>\n",
       "      <td>TAGGGGCTATTAACTTTATTACAACTATTATTAATATGAAAGCCCC...</td>\n",
       "      <td>ENSOCUG00000031196</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1888</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>GL019308</td>\n",
       "      <td>62527</td>\n",
       "      <td>12805</td>\n",
       "      <td>3439</td>\n",
       "      <td>3462</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>88228</td>\n",
       "      <td>AATTATTCATAGCCTCAATGATGAACAAGACATCCAAAAAACAGGA...</td>\n",
       "      <td>AATTATTCACAGCCTcaaTGATGAACAGGATATCCGAAAGATAGGC...</td>\n",
       "      <td>GTAGGAATTATATCCTTCCTCCTCATCGGCTGATGGCATGGCCAAG...</td>\n",
       "      <td>ACGTATACATGAGTTGTTCTTACATAATTGAATATACCTGCAATGC...</td>\n",
       "      <td>TGCCTAGGCGCAATCACTACCTTATTCACAGCCCTATGTGCCCTCA...</td>\n",
       "      <td>GTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGTACAC...</td>\n",
       "      <td>ENSOCUG00000037635</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1824</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>AAGW02081594</td>\n",
       "      <td>14016</td>\n",
       "      <td>6061</td>\n",
       "      <td>3040</td>\n",
       "      <td>3073</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>24576</td>\n",
       "      <td>TTCTTTGGACACCCTGAGGTATATATTCTTATCCTTCCAGGATTCG...</td>\n",
       "      <td>TTTTTCGGGCACCCCGAAGTATATATTCTTATTCTTCCAGGATTTG...</td>\n",
       "      <td>GACATGTCAAAGGAACGCAGAAGCCAGCTTGAAAAAGTCCTCAACA...</td>\n",
       "      <td>TGAAATAGCTATCACTATTCAGGTATGCCACTATTCTTCTAGGCAT...</td>\n",
       "      <td>ATATGAAAGCCCCTGCAATATCTCAATATCAAACCCCCTTATTCGT...</td>\n",
       "      <td>ATTCTCCTAGGTATCTATTTTACCCTACTTCAAGCATCGGAGTATT...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   score  eg2_value  e_value          g_id    g_start  mt_start  g_length  \\\n",
       "0  17245        0.0      0.0            MT          0         0     17245   \n",
       "1   3398        0.0      0.0             1  114731432      6976      4680   \n",
       "2   3120        0.0      0.0      GL019308      67160       883      4892   \n",
       "3   1888        0.0      0.0      GL019308      62527     12805      3439   \n",
       "4   1824        0.0      0.0  AAGW02081594      14016      6061      3040   \n",
       "\n",
       "   mt_length g_strand mt_strand     g_size  \\\n",
       "0      17245        +         +      17245   \n",
       "1       4681        +         +  194850757   \n",
       "2       4945        +         +      88228   \n",
       "3       3462        +         +      88228   \n",
       "4       3073        +         +      24576   \n",
       "\n",
       "                                          g_sequence  \\\n",
       "0  GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...   \n",
       "1  taGTAAAATTATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...   \n",
       "2  CAAGGCCATGAAGCATGCACACACTGCCCATCGCTCTCCTCAAACA...   \n",
       "3  AATTATTCATAGCCTCAATGATGAACAAGACATCCAAAAAACAGGA...   \n",
       "4  TTCTTTGGACACCCTGAGGTATATATTCTTATCCTTCCAGGATTCG...   \n",
       "\n",
       "                                         mt_sequence  \\\n",
       "0  GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...   \n",
       "1  TAGTAAAACCATTACATAACTTTGTCGAAGTTAATTTATAGGTTCA...   \n",
       "2  CAAGGCCATGAAGCACGCACACACCGCCCGTCACCCTCCTCAAGTG...   \n",
       "3  AATTATTCACAGCCTcaaTGATGAACAGGATATCCGAAAGATAGGC...   \n",
       "4  TTTTTCGGGCACCCCGAAGTATATATTCTTATTCTTCCAGGATTTG...   \n",
       "\n",
       "                                       g_up_flanking  \\\n",
       "0                                                NaN   \n",
       "1  ACAAAAGACTCTGAATAGCCAAAGCGATCCTGAACAAGAAAAATCA...   \n",
       "2  GATAAACCCCGATAAACTTTACCACTCTTTGCCAACTCAGCCTCTA...   \n",
       "3  GTAGGAATTATATCCTTCCTCCTCATCGGCTGATGGCATGGCCAAG...   \n",
       "4  GACATGTCAAAGGAACGCAGAAGCCAGCTTGAAAAAGTCCTCAACA...   \n",
       "\n",
       "                                     g_down_flanking  \\\n",
       "0                                                NaN   \n",
       "1  TATATATATTTATATATATATGTATATATGTACATATATATAACTA...   \n",
       "2  ATGAAGGGGCAAAGAGTAATTTTACAGTCGAGAGGGCTAGCAGTCA...   \n",
       "3  ACGTATACATGAGTTGTTCTTACATAATTGAATATACCTGCAATGC...   \n",
       "4  TGAAATAGCTATCACTATTCAGGTATGCCACTATTCTTCTAGGCAT...   \n",
       "\n",
       "  mt_up_flanking (if strand is negative, the coordinates are corrected)  \\\n",
       "0  TACAAATTGTATAATATTTGGACTAAACCCCCTTTCCCACCCCAAG...                      \n",
       "1  CCTTCGCCTCAAAACGAGAAGTAGAGACTATTGAACTAACAACCAC...                      \n",
       "2  TACCGCCATCTTCAGCGAACCCTAAAAAGGAGCAAAAGTAAGCTCA...                      \n",
       "3  TGCCTAGGCGCAATCACTACCTTATTCACAGCCCTATGTGCCCTCA...                      \n",
       "4  ATATGAAAGCCCCTGCAATATCTCAATATCAAACCCCCTTATTCGT...                      \n",
       "\n",
       "  mt_down_flanking (if strand is negative, the coordinates are corrected)  \\\n",
       "0  GTTAATGTAGCTTAACAACAAAGCAAAGCACTGAAAATGCTTAGAT...                        \n",
       "1  AACTGCTAATTCATGCTCCCATGTCTAACCCCATGGCTTTCTCAAC...                        \n",
       "2  TAGGGGCTATTAACTTTATTACAACTATTATTAATATGAAAGCCCC...                        \n",
       "3  GTACACCCGTACGCACGCACGTACACCCGTACGCACGCACGTACAC...                        \n",
       "4  ATTCTCCTAGGTATCTATTTTACCCTACTTCAAGCATCGGAGTATT...                        \n",
       "\n",
       "      ensembl_gene_id                                ensembl_description  \n",
       "0  ENSOCUG00000029081                                                NaN  \n",
       "1  ENSOCUG00000006104  cilia and flagella associated protein 300 [Sou...  \n",
       "2  ENSOCUG00000031196                                                NaN  \n",
       "3  ENSOCUG00000037635                                                NaN  \n",
       "4                 NaN                                                NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "master_array = pd.read_csv(r'Z:/balint/numt/Michael/data/numt_master_array.csv')\n",
    "master_array.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3242"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#genome parts\n",
    "genome_parts = os.listdir(r'Z:\\balint\\numt\\Michael\\data\\genome_parts_in_fasta/')\n",
    "len(genome_parts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['MT', '1', 'GL019308', 'AAGW02081594', 'GL019326', 'GL018741',\n",
       "       '18', '9', '12', 'GL018705', '17', 'GL018700', '19', 'GL019465',\n",
       "       'GL018817', '13', 'GL018826', 'GL018748', '2', '7', '16',\n",
       "       'GL018786', 'GL018706', '4', 'AAGW02081404', 'GL019271',\n",
       "       'GL019056', 'GL019383', 'GL019525', 'GL019149', 'GL018925', '20',\n",
       "       'AAGW02083714', 'GL018992', 'AAGW02079688', 'GL019086', 'GL019010',\n",
       "       '15', 'GL019047', 'GL018862', '11', 'GL019311', '14', '3',\n",
       "       'GL019235', '21', 'GL018744'], dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#genomic identifiers without repetition\n",
    "g_ids = master_array['g_id'].unique()\n",
    "g_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#splitting master array to individual dataframes\n",
    "dfs_dict = {}\n",
    "for g_id in g_ids:\n",
    "    mask = master_array['g_id'] == g_id\n",
    "    dataframe = master_array[mask]\n",
    "    dfs_dict[g_id] = dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>eg2_value</th>\n",
       "      <th>e_value</th>\n",
       "      <th>g_id</th>\n",
       "      <th>g_start</th>\n",
       "      <th>mt_start</th>\n",
       "      <th>g_length</th>\n",
       "      <th>mt_length</th>\n",
       "      <th>g_strand</th>\n",
       "      <th>mt_strand</th>\n",
       "      <th>g_size</th>\n",
       "      <th>g_sequence</th>\n",
       "      <th>mt_sequence</th>\n",
       "      <th>g_up_flanking</th>\n",
       "      <th>g_down_flanking</th>\n",
       "      <th>mt_up_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "      <th>mt_down_flanking (if strand is negative, the coordinates are corrected)</th>\n",
       "      <th>ensembl_gene_id</th>\n",
       "      <th>ensembl_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>73</td>\n",
       "      <td>6.000000e-18</td>\n",
       "      <td>1.100000e-21</td>\n",
       "      <td>4</td>\n",
       "      <td>17873977</td>\n",
       "      <td>3675</td>\n",
       "      <td>197</td>\n",
       "      <td>207</td>\n",
       "      <td>+</td>\n",
       "      <td>+</td>\n",
       "      <td>91394100</td>\n",
       "      <td>ATACTATTAAGTATTCAGCCACACATTTAGAAATATGTCTAACAGA...</td>\n",
       "      <td>ATACTTTCAAGTATCCCTCCTCAGATATAGAAATATGTCTGATAAA...</td>\n",
       "      <td>GCAAGTTCAATTTGTGTACTTTAGGAAAGAAACTATTCCTAAAACA...</td>\n",
       "      <td>GTACAGTTAATTGGTCTGCAGTACTGTTAACTGCTTAGTGCCAGTA...</td>\n",
       "      <td>TAGGCTCATTTCACAGCCACACTAACCCAGAAATATTTACAGTCAA...</td>\n",
       "      <td>TACCCCGAAAATGTTGGTTTATATCCTTCCCGTACTAATTAACCCC...</td>\n",
       "      <td>ENSOCUG00000015079</td>\n",
       "      <td>SEC23 homolog B, coat complex II component [So...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>966</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>13254816</td>\n",
       "      <td>7012</td>\n",
       "      <td>1695</td>\n",
       "      <td>1706</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>GGGATAAGTATAATTGTAGGAATAATAATTTTTAGCATTGTAGGAG...</td>\n",
       "      <td>GGAATAAGTATGATTGTGGGGATAATCGTTTTTAGCATTGTAGAAG...</td>\n",
       "      <td>AGATAGTAGTAGGTGGTAAAGCAAAATTAGTGAGGCTAGCTAGTAG...</td>\n",
       "      <td>AGAAATTACTATTAGGGATGAAGTTCACCCTTTGGTGGAGTGTATT...</td>\n",
       "      <td>ACTAGCCCATTTCCTCCCACAAGGAACCCCTATTCCTCTTATTCCT...</td>\n",
       "      <td>AACCGTATGGTGATCAAAACCCCATATAATTTGAATTAACGCAACA...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>961</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>13256500</td>\n",
       "      <td>9056</td>\n",
       "      <td>1961</td>\n",
       "      <td>1967</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>AATATAATTAAAGAAATTACTATTAGGGATGAAGTTCACCCTTTGG...</td>\n",
       "      <td>AATATGATTAGGGAAATTAGTATTAGAGATCAGGTTCGTCCTTTGG...</td>\n",
       "      <td>ATCGTATGGCTAAGCCTGATGTTATGAAAAGGCCAGATTGCGCTCC...</td>\n",
       "      <td>AGTCCTTTTGGAAAAATTTAAAGCTGAACTAAAATTCTGTTCTGGA...</td>\n",
       "      <td>GGAGGAGGAGACCCTATCCTCTACCAACACCTATTCTGATTTTTCG...</td>\n",
       "      <td>TATTGGCTCAACCAACCTCCTAGGCCTCCTGCCTCACTCATTCACA...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>786</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>4</td>\n",
       "      <td>13253305</td>\n",
       "      <td>4594</td>\n",
       "      <td>1501</td>\n",
       "      <td>1523</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>TGATTGCACAAGGGGCTGTGAAAAGGGTCACAGATGCTCCTCGGCA...</td>\n",
       "      <td>TGAGGGCACATAGGGCTGTGAATAAGGTAGTGATTGCGCCTAGGCA...</td>\n",
       "      <td>CTCCAACCCGCAACCCTCCCCTTATCCACTCCCTCTCCCCTTCCAT...</td>\n",
       "      <td>TCATGAATTAGGGATAAGTATAATTGTAGGAATAATAATTTTTAGC...</td>\n",
       "      <td>GCATTACCATCCTACTCAGCCCTATTACAGACTATATAGCCTACCC...</td>\n",
       "      <td>CACAAAACGACATCAAAAAAATTGTAGCCTTTTCCACATCAAGTCA...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>572</td>\n",
       "      <td>1.800000e-255</td>\n",
       "      <td>3.100000e-259</td>\n",
       "      <td>4</td>\n",
       "      <td>13259345</td>\n",
       "      <td>16374</td>\n",
       "      <td>998</td>\n",
       "      <td>994</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>CTCTATTCTTAATTTACTTCTAAATCCTCCTTTAGCCCTTAACATT...</td>\n",
       "      <td>CTCTATTCTTAATTTACTACTAAATCCTCCTTTGGCCCTTAGAGTT...</td>\n",
       "      <td>TAGTGAGAGAGAGAGAGAGAGAAAGAGACAGAGAGAAAGGTCTTTC...</td>\n",
       "      <td>GCATATAAATAAAAATTAAAATTTCTAGTTCATTAAGCCTAAACAT...</td>\n",
       "      <td>CCTTTCCCACCCCAAGTCAGACAGCTCAGGGCATCTAAATTTTGAA...</td>\n",
       "      <td>TGCTTAATTGAACAAGGCCATGAAGCACGCACACACCGCCCGTCAC...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>556</td>\n",
       "      <td>7.500000e-248</td>\n",
       "      <td>1.300000e-251</td>\n",
       "      <td>4</td>\n",
       "      <td>13259345</td>\n",
       "      <td>33619</td>\n",
       "      <td>866</td>\n",
       "      <td>871</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>CTCTATTCTTAATTTACTTCTAAATCCTCCTTTAGCCCTTAACATT...</td>\n",
       "      <td>CTCTATTCTTAATTTACTACTAAATCCTCCTTTGGCCCTTAGAGTT...</td>\n",
       "      <td>TAGTGAGAGAGAGAGAGAGAGAAAGAGACAGAGAGAAAGGTCTTTC...</td>\n",
       "      <td>CATATATGTAAGGTTTGGCTAAAAGAATTTCTAGTTAAAATCCTAA...</td>\n",
       "      <td>TACAAATTGTATAATATTTGGACTAAACCCCCTTTCCCACCCCAAG...</td>\n",
       "      <td>TGCTTAATTGAACAAGGCCATGAAGCACGCACACACCGCCCGTCAC...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>401</td>\n",
       "      <td>4.500000e-174</td>\n",
       "      <td>7.900000e-178</td>\n",
       "      <td>4</td>\n",
       "      <td>13258434</td>\n",
       "      <td>15725</td>\n",
       "      <td>680</td>\n",
       "      <td>682</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>ACATCTAAATTTTCATTAAGTTTATGAAGTCCTTTTGGAAAAATTT...</td>\n",
       "      <td>ACATTTAAATTTTCATTAGGGTTTT-ATGTCCTTTAGGAAAA-TTT...</td>\n",
       "      <td>GAGAAGAAGTAGCTAAGACAATTCCTGTGAGCCTACCTACAGTGAA...</td>\n",
       "      <td>TGTTTGTTTGTTTGACAGGTAGAGTTTTAGATAGTGAGAGAGAGAG...</td>\n",
       "      <td>ATAAACCCCGATAAACCCTACCACTCTTTGCCAACTCAGCCTATAT...</td>\n",
       "      <td>TACTCTAAAGAGGGACAGCTCTTTAGATATAGGATACAGCCTTTTA...</td>\n",
       "      <td>ENSOCUG00000008071</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>210</td>\n",
       "      <td>3.700000e-83</td>\n",
       "      <td>6.500000e-87</td>\n",
       "      <td>4</td>\n",
       "      <td>35271214</td>\n",
       "      <td>3810</td>\n",
       "      <td>1173</td>\n",
       "      <td>1214</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>91394100</td>\n",
       "      <td>ATATCTAGTGAAAATGATAGTGTACTGTGGCTTGTTCAAAGATTCA...</td>\n",
       "      <td>ATATCTAGGAGTAATGTGGCTATGTTTTGGCTAGCTGAAAGGTTTA...</td>\n",
       "      <td>GGTCTCCACCATCCACATGGGAGACTTGGATTGGCTTTATGGCTCC...</td>\n",
       "      <td>GGGGTGGGCCTTCAGCGCAGCAGTTAAGTTTCCAACTGGGACACCC...</td>\n",
       "      <td>ACTTCTCTATACTATTCACCCCAATTGCACTATTTGTTACATGGTC...</td>\n",
       "      <td>GACCTGAACTGAAAAAGCGATCCCGAAAAATATCTCAGACATTCAA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    score      eg2_value        e_value g_id   g_start  mt_start  g_length  \\\n",
       "64     73   6.000000e-18   1.100000e-21    4  17873977      3675       197   \n",
       "74    966   0.000000e+00   0.000000e+00    4  13254816      7012      1695   \n",
       "75    961   0.000000e+00   0.000000e+00    4  13256500      9056      1961   \n",
       "76    786   0.000000e+00   0.000000e+00    4  13253305      4594      1501   \n",
       "80    572  1.800000e-255  3.100000e-259    4  13259345     16374       998   \n",
       "82    556  7.500000e-248  1.300000e-251    4  13259345     33619       866   \n",
       "86    401  4.500000e-174  7.900000e-178    4  13258434     15725       680   \n",
       "91    210   3.700000e-83   6.500000e-87    4  35271214      3810      1173   \n",
       "\n",
       "    mt_length g_strand mt_strand    g_size  \\\n",
       "64        207        +         +  91394100   \n",
       "74       1706        +         -  91394100   \n",
       "75       1967        +         -  91394100   \n",
       "76       1523        +         -  91394100   \n",
       "80        994        +         -  91394100   \n",
       "82        871        +         -  91394100   \n",
       "86        682        +         -  91394100   \n",
       "91       1214        +         -  91394100   \n",
       "\n",
       "                                           g_sequence  \\\n",
       "64  ATACTATTAAGTATTCAGCCACACATTTAGAAATATGTCTAACAGA...   \n",
       "74  GGGATAAGTATAATTGTAGGAATAATAATTTTTAGCATTGTAGGAG...   \n",
       "75  AATATAATTAAAGAAATTACTATTAGGGATGAAGTTCACCCTTTGG...   \n",
       "76  TGATTGCACAAGGGGCTGTGAAAAGGGTCACAGATGCTCCTCGGCA...   \n",
       "80  CTCTATTCTTAATTTACTTCTAAATCCTCCTTTAGCCCTTAACATT...   \n",
       "82  CTCTATTCTTAATTTACTTCTAAATCCTCCTTTAGCCCTTAACATT...   \n",
       "86  ACATCTAAATTTTCATTAAGTTTATGAAGTCCTTTTGGAAAAATTT...   \n",
       "91  ATATCTAGTGAAAATGATAGTGTACTGTGGCTTGTTCAAAGATTCA...   \n",
       "\n",
       "                                          mt_sequence  \\\n",
       "64  ATACTTTCAAGTATCCCTCCTCAGATATAGAAATATGTCTGATAAA...   \n",
       "74  GGAATAAGTATGATTGTGGGGATAATCGTTTTTAGCATTGTAGAAG...   \n",
       "75  AATATGATTAGGGAAATTAGTATTAGAGATCAGGTTCGTCCTTTGG...   \n",
       "76  TGAGGGCACATAGGGCTGTGAATAAGGTAGTGATTGCGCCTAGGCA...   \n",
       "80  CTCTATTCTTAATTTACTACTAAATCCTCCTTTGGCCCTTAGAGTT...   \n",
       "82  CTCTATTCTTAATTTACTACTAAATCCTCCTTTGGCCCTTAGAGTT...   \n",
       "86  ACATTTAAATTTTCATTAGGGTTTT-ATGTCCTTTAGGAAAA-TTT...   \n",
       "91  ATATCTAGGAGTAATGTGGCTATGTTTTGGCTAGCTGAAAGGTTTA...   \n",
       "\n",
       "                                        g_up_flanking  \\\n",
       "64  GCAAGTTCAATTTGTGTACTTTAGGAAAGAAACTATTCCTAAAACA...   \n",
       "74  AGATAGTAGTAGGTGGTAAAGCAAAATTAGTGAGGCTAGCTAGTAG...   \n",
       "75  ATCGTATGGCTAAGCCTGATGTTATGAAAAGGCCAGATTGCGCTCC...   \n",
       "76  CTCCAACCCGCAACCCTCCCCTTATCCACTCCCTCTCCCCTTCCAT...   \n",
       "80  TAGTGAGAGAGAGAGAGAGAGAAAGAGACAGAGAGAAAGGTCTTTC...   \n",
       "82  TAGTGAGAGAGAGAGAGAGAGAAAGAGACAGAGAGAAAGGTCTTTC...   \n",
       "86  GAGAAGAAGTAGCTAAGACAATTCCTGTGAGCCTACCTACAGTGAA...   \n",
       "91  GGTCTCCACCATCCACATGGGAGACTTGGATTGGCTTTATGGCTCC...   \n",
       "\n",
       "                                      g_down_flanking  \\\n",
       "64  GTACAGTTAATTGGTCTGCAGTACTGTTAACTGCTTAGTGCCAGTA...   \n",
       "74  AGAAATTACTATTAGGGATGAAGTTCACCCTTTGGTGGAGTGTATT...   \n",
       "75  AGTCCTTTTGGAAAAATTTAAAGCTGAACTAAAATTCTGTTCTGGA...   \n",
       "76  TCATGAATTAGGGATAAGTATAATTGTAGGAATAATAATTTTTAGC...   \n",
       "80  GCATATAAATAAAAATTAAAATTTCTAGTTCATTAAGCCTAAACAT...   \n",
       "82  CATATATGTAAGGTTTGGCTAAAAGAATTTCTAGTTAAAATCCTAA...   \n",
       "86  TGTTTGTTTGTTTGACAGGTAGAGTTTTAGATAGTGAGAGAGAGAG...   \n",
       "91  GGGGTGGGCCTTCAGCGCAGCAGTTAAGTTTCCAACTGGGACACCC...   \n",
       "\n",
       "   mt_up_flanking (if strand is negative, the coordinates are corrected)  \\\n",
       "64  TAGGCTCATTTCACAGCCACACTAACCCAGAAATATTTACAGTCAA...                      \n",
       "74  ACTAGCCCATTTCCTCCCACAAGGAACCCCTATTCCTCTTATTCCT...                      \n",
       "75  GGAGGAGGAGACCCTATCCTCTACCAACACCTATTCTGATTTTTCG...                      \n",
       "76  GCATTACCATCCTACTCAGCCCTATTACAGACTATATAGCCTACCC...                      \n",
       "80  CCTTTCCCACCCCAAGTCAGACAGCTCAGGGCATCTAAATTTTGAA...                      \n",
       "82  TACAAATTGTATAATATTTGGACTAAACCCCCTTTCCCACCCCAAG...                      \n",
       "86  ATAAACCCCGATAAACCCTACCACTCTTTGCCAACTCAGCCTATAT...                      \n",
       "91  ACTTCTCTATACTATTCACCCCAATTGCACTATTTGTTACATGGTC...                      \n",
       "\n",
       "   mt_down_flanking (if strand is negative, the coordinates are corrected)  \\\n",
       "64  TACCCCGAAAATGTTGGTTTATATCCTTCCCGTACTAATTAACCCC...                        \n",
       "74  AACCGTATGGTGATCAAAACCCCATATAATTTGAATTAACGCAACA...                        \n",
       "75  TATTGGCTCAACCAACCTCCTAGGCCTCCTGCCTCACTCATTCACA...                        \n",
       "76  CACAAAACGACATCAAAAAAATTGTAGCCTTTTCCACATCAAGTCA...                        \n",
       "80  TGCTTAATTGAACAAGGCCATGAAGCACGCACACACCGCCCGTCAC...                        \n",
       "82  TGCTTAATTGAACAAGGCCATGAAGCACGCACACACCGCCCGTCAC...                        \n",
       "86  TACTCTAAAGAGGGACAGCTCTTTAGATATAGGATACAGCCTTTTA...                        \n",
       "91  GACCTGAACTGAAAAAGCGATCCCGAAAAATATCTCAGACATTCAA...                        \n",
       "\n",
       "       ensembl_gene_id                                ensembl_description  \n",
       "64  ENSOCUG00000015079  SEC23 homolog B, coat complex II component [So...  \n",
       "74  ENSOCUG00000008071                                                NaN  \n",
       "75  ENSOCUG00000008071                                                NaN  \n",
       "76  ENSOCUG00000008071                                                NaN  \n",
       "80  ENSOCUG00000008071                                                NaN  \n",
       "82  ENSOCUG00000008071                                                NaN  \n",
       "86  ENSOCUG00000008071                                                NaN  \n",
       "91                 NaN                                                NaN  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs_dict['4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "starts = dfs_dict['4']['g_start'].tolist()\n",
    "lengths = dfs_dict['4']['g_length'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "ends = []\n",
    "for start, length in zip(starts, lengths):\n",
    "    ends.append(start + length)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "ends.sort(reverse = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
